{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Neural and Evolutionary Learning - Final Project"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Notebook 2 - Problem Applications\n",
    "\n",
    "## **2.4.** *NeuroEvolution of Augmenting Topologies* - `NEAT`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "|       Student Name      |    Student ID   |\n",
    "|           ---           |       ---       |\n",
    "|     Andriani Kakoulli   |     20230484    |\n",
    "|       Susana Pires      |     20230540    |"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **1. Introduction**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> **CONTEXT** <br>\n",
    "\n",
    "Data from Automatic Milking Systems (AMSs). <br>\n",
    "\n",
    "Milk contents (fat, protein, lactose) are emasured at each milking event, to evaluate milk quality. This measurement is through a colourimetric mehtod that requires calibration every 2 weeks. <br>\n",
    "\n",
    "> **GOAL** <br>\n",
    "\n",
    "To infer the milk components - *Lactose* - from the other sensor data from the milking robots. <br>\n",
    "Develop model to **predict** Lactose content."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **2. Environment preparation**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **2.1. Imports**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %pip install git+https://github.com/CodeReclaimers/neat-python.git"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import csv\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "from sklearn.metrics import mean_squared_error\n",
    "import statistics as sts\n",
    "from torch.utils.data import DataLoader, TensorDataset\n",
    "\n",
    "import torch\n",
    "from torch.utils.data import TensorDataset, DataLoader\n",
    "\n",
    "#import neat-python\n",
    "import neat\n",
    "import graphviz\n",
    "from NEAT.visualize import draw_net, plot_stats, plot_species"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **2.2. Global settings**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "seed = 91\n",
    "\n",
    "# seeds = 6, 8, 11, 12, 21, 27, 29, 33, 66, 91\n",
    "# Susana - my device will always be cpu because my pc doesn't have a NVIDIA (nor a recent AMD) GPU.\n",
    "device = 'cuda' if torch.cuda.is_available() else 'cpu'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **2.3. Data**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* **Import data**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(177, 13), (177, 1)]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = pd.read_csv('dt_cleaned.csv').to_numpy()[:, 1:]\n",
    "target = pd.read_csv('y_lact_cleaned.csv').to_numpy()\n",
    "[data.shape, target.shape]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* **Data split** <br>\n",
    "Since we are working with a very small dataset, the best decision for the split is by K-fold Cross Validation, but first we will use the Hold-Out Method. Then, after everything is running smoothly, we might adapt to Cross Validation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Train+Val and Test sets\n",
    "X_tval, X_test, y_tval, y_test = train_test_split(data, target)\n",
    "\n",
    "# Train and Val sets (load from X_tval and y_tval)\n",
    "X_train, X_val, y_train, y_val = train_test_split(X_tval, y_tval)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(99, 13), (99, 1), (33, 13), (33, 1), (45, 13), (45, 1)]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[X_train.shape, y_train.shape, X_val.shape, y_val.shape, X_test.shape, y_test.shape]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **3. NEAT Configuration**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **3.1. Defining function `eval_genomes`:** <br>\n",
    "\n",
    "Get the fitness (in this case RMSE) for each train and validation set."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def eval_genomes(genomes, config):\n",
    "    for genome_id, genome in genomes:\n",
    "        net = neat.nn.FeedForwardNetwork.create(genome, config)\n",
    "        \n",
    "        # Negative RMSE\n",
    "        genome.fitness = 0.0\n",
    "        for xi, xo in zip(X_train, y_train):\n",
    "            output = net.activate(xi)                           # activate is a method of FeedForwardNetwork that throws the input through the network\n",
    "            genome.fitness += (output[0] - xo)**2\n",
    "        genome.fitness = -(genome.fitness[0]/y_train.shape[0])**.5\n",
    "\n",
    "        # Negative RMSE Validation\n",
    "        genome.fitness_val = 0.0\n",
    "        for xi, xo in zip(X_val, y_val):\n",
    "            output = net.activate(xi)\n",
    "            genome.fitness_val += (output[0] - xo)**2\n",
    "        genome.fitness_val = -(genome.fitness_val[0]/y_val.shape[0])**.5\n",
    "\n",
    "        # TEST\n",
    "        # Negative RMSE Test\n",
    "        genome.fitness_test = 0.0\n",
    "        for xi, xo in zip(X_test, y_test):\n",
    "            output = net.activate(xi)\n",
    "            genome.fitness_test += (output[0] - xo)**2\n",
    "        genome.fitness_test = -(genome.fitness_test[0]/y_test.shape[0])**.5"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **3.2. Set *configuration file*** <br>\n",
    "\n",
    "Adapted configuration file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#--- parameters for the NEAT Milk Lactose experiment ---#\n",
      "\n",
      "[NEAT]\n",
      "# The function used to compute the termination \n",
      "# criterion from the set of genome fitnesses. \n",
      "fitness_criterion     = max\n",
      "\n",
      "# When the fitness computed by fitness_criterion \n",
      "# meets or exceeds this threshold, the evolution \n",
      "# process will terminate, with a call to any \n",
      "# registered reporting class 'found_solution' method.\n",
      "# Note\n",
      "# The found_solution method is not called if the maximum \n",
      "# number of generations is reached without the above \n",
      "# threshold being passed.\n",
      "fitness_threshold     = 0.\n",
      "\n",
      "# If no_fitness_termination evaluates to True, \n",
      "# then the fitness_criterion and fitness_threshold are ignored \n",
      "# for termination; only valid if termination by a maximum number \n",
      "# of generations passed to population.Population.run() is enabled, \n",
      "# and the found_solution method is called upon \n",
      "# generation number termination.\n",
      "no_fitness_termination = False\n",
      "\n",
      "# The number of individuals in NEAT population.\n",
      "pop_size              = 250\n",
      "\n",
      "# If this evaluates to True, when all species simultaneously \n",
      "# become extinct due to stagnation, a new random population will be created. \n",
      "# If False, a CompleteExtinctionException will be thrown.\n",
      "reset_on_extinction   = False\n",
      "\n",
      "# The DefaultGenome section specifies parameters \n",
      "# for the builtin DefaultGenome class. This section \n",
      "# is only necessary if you specify this class as the genome \n",
      "# implementation when creating the Config instance; \n",
      "# otherwise you need to include whatever configuration \n",
      "# (if any) is required for your particular implementation.\n",
      "[DefaultGenome]\n",
      "\n",
      "# node activation options\n",
      "\n",
      "# The default activation function attribute assigned to \n",
      "# new nodes. If none is given, or â€œrandomâ€ is specified, \n",
      "# one of the activation_options will be chosen at random.\n",
      "activation_default      = identity\n",
      "\n",
      "# The probability that mutation will replace the nodeâ€™s activation \n",
      "# function with a randomly-determined member of the activation_options. \n",
      "# Valid values are in [0.0, 1.0].\n",
      "activation_mutate_rate  = 0.5\n",
      "\n",
      "# A space-separated list of the activation functions that \n",
      "# may be used by nodes. This defaults to sigmoid. \n",
      "# The built-in available functions can be found in Overview \n",
      "# of builtin activation functions; \n",
      "# more can be added as described in Customizing Behavior.\n",
      "activation_options      = sigmoid identity gauss relu\n",
      "\n",
      "# node aggregation options\n",
      "\n",
      "# The default aggregation function attribute assigned to \n",
      "# new nodes. If none is given, or â€œrandomâ€ is specified, \n",
      "# one of the aggregation_options will be chosen at random.\n",
      "aggregation_default     = sum\n",
      "\n",
      "# The probability that mutation will replace the nodeâ€™s \n",
      "# aggregation function with a randomly-determined member \n",
      "# of the aggregation_options. Valid values are in [0.0, 1.0].\n",
      "aggregation_mutate_rate = 0.5\n",
      "\n",
      "# A space-separated list of the aggregation functions that \n",
      "# may be used by nodes. This defaults to â€œsumâ€. The available \n",
      "# functions (defined in aggregations) are: sum, product, min, \n",
      "# max, mean, median, and maxabs (which returns the input value \n",
      "# with the greatest absolute value; the returned value may \n",
      "# be positive or negative). New aggregation functions can be \n",
      "# defined similarly to new activation functions. \n",
      "aggregation_options     = mean sum\n",
      "\n",
      "# node bias options\n",
      "\n",
      "# The mean and standard deviation of the normal/gaussian \n",
      "# distribution, if it is used to select bias attribute \n",
      "# values for new nodes.\n",
      "bias_init_mean          = 0.0\n",
      "bias_init_stdev         = 1.0\n",
      "\n",
      "# bias_init_type\n",
      "# If set to gaussian or normal, then the initialization is \n",
      "# to a normal/gaussian distribution. If set to uniform, \n",
      "# a uniform distribution from \n",
      "# max(ğ‘ğ‘–ğ‘ğ‘ _ğ‘šğ‘–ğ‘›_ğ‘£ğ‘ğ‘™ğ‘¢ğ‘’,(ğ‘ğ‘–ğ‘ğ‘ _ğ‘–ğ‘›ğ‘–ğ‘¡_ğ‘šğ‘’ğ‘ğ‘›âˆ’(ğ‘ğ‘–ğ‘ğ‘ _ğ‘–ğ‘›ğ‘–ğ‘¡_ğ‘ ğ‘¡ğ‘‘ğ‘’ğ‘£âˆ—2))) to \n",
      "# min(ğ‘ğ‘–ğ‘ğ‘ _ğ‘šğ‘ğ‘¥_ğ‘£ğ‘ğ‘™ğ‘¢ğ‘’,(ğ‘ğ‘–ğ‘ğ‘ _ğ‘–ğ‘›ğ‘–ğ‘¡_ğ‘šğ‘’ğ‘ğ‘›+(ğ‘ğ‘–ğ‘ğ‘ _ğ‘–ğ‘›ğ‘–ğ‘¡_ğ‘ ğ‘¡ğ‘‘ğ‘’ğ‘£âˆ—2))). \n",
      "# Note that the standard deviation of a uniform distribution \n",
      "# is not range/0.25, as implied by this, but the range divided \n",
      "# by a bit over 0.288 (the square root of 12); however, this \n",
      "# approximation makes setting the range much easier.) \n",
      "# This defaults to â€œgaussianâ€.\n",
      "\n",
      "# The maximum and minimum allowed bias value. Biases above \n",
      "# this value will be clamped to this value.\n",
      "bias_max_value          = 30.0\n",
      "bias_min_value          = -30.0\n",
      "\n",
      "# The standard deviation of the zero-centered normal/gaussian \n",
      "# distribution from which a bias value mutation is drawn.\n",
      "bias_mutate_power       = 0.5\n",
      "\n",
      "# The probability that mutation will change the bias of a \n",
      "# node by adding a random value.\n",
      "bias_mutate_rate        = 0.7\n",
      "\n",
      "# The probability that mutation will replace the bias of a \n",
      "# node with a newly chosen random value (as if it were a new node).\n",
      "bias_replace_rate       = 0.1\n",
      "\n",
      "# genome compatibility options\n",
      "\n",
      "# The coefficient for the disjoint and excess gene \n",
      "# countsâ€™ contribution to the genomic distance.\n",
      "compatibility_disjoint_coefficient = 1.0\n",
      "\n",
      "# The coefficient for each weight, bias, or response \n",
      "# multiplier differenceâ€™s contribution to the genomic \n",
      "# distance (for homologous nodes or connections). \n",
      "# This is also used as the value to add for differences \n",
      "# in activation functions, aggregation functions, or \n",
      "# enabled/disabled status.\n",
      "compatibility_weight_coefficient   = 0.5\n",
      "\n",
      "# connection add/remove rates\n",
      "\n",
      "# The probability that mutation will add or delete \n",
      "# a connection between existing nodes. \n",
      "# Valid values are in [0.0, 1.0].\n",
      "conn_add_prob           = 0.5\n",
      "conn_delete_prob        = 0.5\n",
      "\n",
      "# connection enable options\n",
      "\n",
      "# The default enabled attribute of newly created \n",
      "# connections. Valid values are True and False.\n",
      "# Note: â€œNewly created connectionsâ€ include ones in \n",
      "# newly-created genomes, if those have initial connections \n",
      "# (from the setting of the initial_connection variable).\n",
      "enabled_default         = True\n",
      "\n",
      "# The probability that mutation will replace (50/50\n",
      "# chance of True or False) the enabled status of \n",
      "# a connection. Valid values are in [0.0, 1.0].\n",
      "enabled_mutate_rate     = 0.01\n",
      "\n",
      "# If this evaluates to True, generated networks will not \n",
      "# be allowed to have recurrent connections \n",
      "# (they will be feedforward). Otherwise they may be \n",
      "# (but are not forced to be) recurrent.\n",
      "feed_forward            = True\n",
      "\n",
      "# Specifies the initial connectivity of newly-created \n",
      "# genomes. (Note the effects on settings other than \n",
      "# unconnected of the enabled_default parameter.) \n",
      "# There are seven allowed values:\n",
      "#   unconnected - No connections are initially present. \n",
      "#   This is the default.\n",
      "#   fs_neat_nohidden - One randomly-chosen input node has \n",
      "#   one connection to each output node. (This is one version \n",
      "#   of the FS-NEAT scheme; â€œFSâ€ stands for â€œFeature Selectionâ€.)\n",
      "#   fs_neat_hidden - One randomly-chosen input node has one \n",
      "#   connection to each hidden and output node. (This is \n",
      "#   another version of the FS-NEAT scheme. \n",
      "#   If there are no hidden nodes, it is the same as fs_neat_nohidden.)\n",
      "#   full_nodirect - Each input node is connected to all hidden \n",
      "#   nodes, if there are any, and each hidden node is connected \n",
      "#   to all output nodes; otherwise, each input node is connected \n",
      "#   to all output nodes. Genomes with feed_forward set to F\n",
      "#   alse will also have recurrent (loopback, in this case) \n",
      "#   connections from each hidden or output node to itself.\n",
      "#   full_direct - Each input node is connected to all hidden \n",
      "#   and output nodes, and each hidden node is connected to all \n",
      "#   output nodes. Genomes with feed_forward set to False will \n",
      "#   also have recurrent (loopback, in this case) connections from \n",
      "#   each hidden or output node to itself.\n",
      "initial_connection      = full_nodirect\n",
      "\n",
      "# node add/remove rates\n",
      "\n",
      "# The probability that mutation will add a new node \n",
      "# (essentially replacing an existing connection, the enabled \n",
      "# status of which will be set to False). \n",
      "# Valid values are in [0.0, 1.0].\n",
      "node_add_prob           = 0.5\n",
      "node_delete_prob        = 0.5\n",
      "\n",
      "# network parameters\n",
      "\n",
      "# The number of hidden nodes to add to each \n",
      "# genome in the initial population. Note that\n",
      "# These are the nodes other than input nodes and \n",
      "# output nodes. In the original NEAT (NeuroEvolution of \n",
      "# Augmenting Topologies) algorithm, networks start \n",
      "# with no hidden nodes, and evolve more \n",
      "# complexity as necessary - thus â€œAugmenting Topologiesâ€.\n",
      "num_hidden              = 5\n",
      "\n",
      "# The number of input nodes, through which the \n",
      "# network receives inputs.\n",
      "# ----- Change according to number of features -----\n",
      "num_inputs              = 13\n",
      "\n",
      "# The number of output nodes, to which the network \n",
      "# delivers outputs.\n",
      "num_outputs             = 1\n",
      "\n",
      "# node response options\n",
      "\n",
      "# The mean and standard deviation of the normal/gaussian \n",
      "# distribution, if it is used to select response multiplier \n",
      "# attribute values for new nodes.\n",
      "response_init_mean      = 1.0\n",
      "response_init_stdev     = 0.0\n",
      "# response_init_type: If set to gaussian or normal, then the \n",
      "# initialization is to a normal/gaussian distribution. \n",
      "\n",
      "# The maximum/minimum allowed response multiplier. \n",
      "# Response multipliers above this value will be clamped to this value.\n",
      "response_max_value      = 30.0\n",
      "response_min_value      = -30.0\n",
      "\n",
      "# The standard deviation of the zero-centered normal/gaussian \n",
      "# distribution from which a response multiplier mutation is drawn.\n",
      "response_mutate_power   = 0.0\n",
      "\n",
      "# The probability that mutation will change the response \n",
      "# multiplier of a node by adding a random value.\n",
      "response_mutate_rate    = 0.0\n",
      "\n",
      "# The probability that mutation will replace the response \n",
      "# multiplier of a node with a newly chosen random value \n",
      "# (as if it were a new node).\n",
      "response_replace_rate   = 0.0\n",
      "\n",
      "# single_structural_mutation: If this evaluates to True, only \n",
      "# one structural mutation (the addition or removal of a node \n",
      "# or connection) will be allowed per genome per generation.\n",
      "# New in version 0.92.\n",
      "\n",
      "# structural_mutation_surer: If this evaluates to True, \n",
      "# then an attempt to add a node to a genome lacking connections \n",
      "# will result in adding a connection instead; furthermore, \n",
      "# if an attempt to add a connection tries to add a connection \n",
      "# that already exists, that connection will be enabled. \n",
      "# If this is set to default, then it acts as if it had the \n",
      "# same value as single_structural_mutation (above). \n",
      "# This defaults to â€œdefaultâ€. New in version 0.92.\n",
      "\n",
      "# connection weight options\n",
      "\n",
      "# The mean and standard deviation of the normal/gaussian \n",
      "# distribution used to select weight attribute values \n",
      "# for new connections.\n",
      "weight_init_mean        = 0.0\n",
      "weight_init_stdev       = 1.0\n",
      "\n",
      "# weight_init_type: If set to gaussian or normal, \n",
      "# then the initialization is to a normal/gaussian distribution.\n",
      "# New in version 0.92.\n",
      "\n",
      "# The maximum/minimum allowed weight value. Weights \n",
      "# above/below this value will be clamped to this value.\n",
      "weight_max_value        = 30\n",
      "weight_min_value        = -30\n",
      "\n",
      "# The standard deviation of the zero-centered normal/gaussian \n",
      "# distribution from which a weight value mutation is drawn.\n",
      "weight_mutate_power     = 0.5\n",
      "\n",
      "# The probability that mutation will change the weight \n",
      "# of a connection by adding a random value.\n",
      "weight_mutate_rate      = 0.7\n",
      "\n",
      "# The probability that mutation will replace the weight of a \n",
      "# connection with a newly chosen random value (as if it were a new connection).\n",
      "weight_replace_rate     = 0.1\n",
      "\n",
      "# The DefaultSpeciesSet section specifies parameters \n",
      "# for the builtin DefaultSpeciesSet class. This section \n",
      "# is only necessary if you specify this class as the genome \n",
      "# implementation when creating the Config instance; \n",
      "# otherwise you need to include whatever configuration \n",
      "# (if any) is required for your particular implementation.\n",
      "[DefaultSpeciesSet]\n",
      "\n",
      "# Individuals whose genomic distance is less than \n",
      "# this threshold are considered to be in the same species.\n",
      "compatibility_threshold = 5.\n",
      "\n",
      "# The DefaultStagnation section specifies parameters \n",
      "# for the builtin DefaultStagnation class. This section \n",
      "# is only necessary if you specify this class as the \n",
      "# stagnation implementation when creating the Config instance; \n",
      "# otherwise you need to include whatever configuration \n",
      "# (if any) is required for your particular implementation.\n",
      "[DefaultStagnation]\n",
      "\n",
      "# The function used to compute species fitness.\n",
      "# This defaults to ``mean``. \n",
      "# Allowed values are: max, min, mean, and median\n",
      "# Note: This is not used for calculating species \n",
      "# fitness for apportioning reproduction (which always uses mean).\n",
      "species_fitness_func = max\n",
      "\n",
      "# Species that have not shown improvement in more than \n",
      "# this number of generations will be considered stagnant \n",
      "# and removed. This defaults to 15.\n",
      "max_stagnation       = 20\n",
      "\n",
      "# The number of species that will be protected from stagnation; \n",
      "# mainly intended to prevent total extinctions caused by \n",
      "# all species becoming stagnant before new species arise. \n",
      "# For example, a species_elitism setting of 3 will prevent \n",
      "# the 3 species with the highest species fitness from \n",
      "# being removed for stagnation regardless of the amount \n",
      "# of time they have not shown improvement. This defaults to 0.\n",
      "species_elitism      = 2\n",
      "\n",
      "# The DefaultReproduction section specifies parameters for the \n",
      "# builtin DefaultReproduction class.\n",
      "# This section is only necessary if you specify this class \n",
      "# as the reproduction implementation when creating the Config instance; \n",
      "# otherwise you need to include whatever configuration \n",
      "# (if any) is required for your particular implementation.\n",
      "[DefaultReproduction]\n",
      "\n",
      "# The number of most-fit individuals in each species that will \n",
      "# be preserved as-is from one generation to the next. This defaults to 0.\n",
      "elitism            = 2\n",
      "\n",
      "# The fraction for each species allowed to reproduce \n",
      "# each generation. This defaults to 0.2.\n",
      "survival_threshold = 0.2\n",
      "\n",
      "# The minimum number of genomes per species after \n",
      "# reproduction. This defaults to 2.\n",
      "# min_species_size\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "config_file = 'NEAT\\config-feedforward-neat'\n",
    "with open(config_file, 'r') as f:\n",
    "    print(f.read())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **3.3. Load *configuration file***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "config = neat.Config(neat.DefaultGenome, neat.DefaultReproduction,\n",
    "                     neat.DefaultSpeciesSet, neat.DefaultStagnation,\n",
    "                     config_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **4. Experiment**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **4.1. Create a *population*** <br>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It is the top-level object for a NEAT run, it's based on the loaded configuration *config*."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "pp = neat.Population(config)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Set some ways of checking the progress."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "pp.add_reporter(neat.StdOutReporter(False))\n",
    "statistics = neat.StatisticsReporter()\n",
    "pp.add_reporter(statistics)\n",
    "pp.add_reporter(neat.Checkpointer(5))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **4.2. Run until finding a solution.** <br>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " ****** Running generation 0 ****** \n",
      "\n",
      "Population's average fitness: -999.09091 stdev: 864.20897\n",
      "Best fitness: -50.61887 - size: (6, 70) - species 1 - id 19\n",
      "Average adjusted fitness: 0.864\n",
      "Mean genetic distance 3.557, standard deviation 0.286\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 3.525 sec\n",
      "\n",
      " ****** Running generation 1 ****** \n",
      "\n",
      "Population's average fitness: -295.31114 stdev: 457.24555\n",
      "Best fitness: -2.97723 - size: (6, 55) - species 1 - id 364\n",
      "Average adjusted fitness: 0.924\n",
      "Mean genetic distance 3.497, standard deviation 0.290\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 3.118 sec (3.322 average)\n",
      "\n",
      " ****** Running generation 2 ****** \n",
      "\n",
      "Population's average fitness: -85.32011 stdev: 267.57821\n",
      "Best fitness: -1.53926 - size: (6, 56) - species 1 - id 646\n",
      "Average adjusted fitness: 0.960\n",
      "Mean genetic distance 3.086, standard deviation 0.444\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 2.917 sec (3.187 average)\n",
      "\n",
      " ****** Running generation 3 ****** \n",
      "\n",
      "Population's average fitness: -67.98615 stdev: 304.52606\n",
      "Best fitness: -1.53926 - size: (6, 56) - species 1 - id 646\n",
      "Average adjusted fitness: 0.978\n",
      "Mean genetic distance 2.856, standard deviation 0.507\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 3.264 sec (3.206 average)\n",
      "\n",
      " ****** Running generation 4 ****** \n",
      "\n",
      "Population's average fitness: -66.41463 stdev: 206.70357\n",
      "Best fitness: -1.45565 - size: (6, 43) - species 1 - id 1106\n",
      "Average adjusted fitness: 0.959\n",
      "Mean genetic distance 3.063, standard deviation 0.442\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 2.719 sec (3.109 average)\n",
      "Saving checkpoint to neat-checkpoint-4\n",
      "\n",
      " ****** Running generation 5 ****** \n",
      "\n",
      "Population's average fitness: -74.01210 stdev: 247.82662\n",
      "Best fitness: -0.70867 - size: (7, 69) - species 1 - id 1385\n",
      "Average adjusted fitness: 0.965\n",
      "Mean genetic distance 2.852, standard deviation 0.454\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 2.695 sec (3.040 average)\n",
      "\n",
      " ****** Running generation 6 ****** \n",
      "\n",
      "Population's average fitness: -60.33345 stdev: 286.55898\n",
      "Best fitness: -0.70867 - size: (7, 69) - species 1 - id 1385\n",
      "Average adjusted fitness: 0.985\n",
      "Mean genetic distance 2.685, standard deviation 0.387\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 2.300 sec (2.934 average)\n",
      "\n",
      " ****** Running generation 7 ****** \n",
      "\n",
      "Population's average fitness: -35.67334 stdev: 121.95392\n",
      "Best fitness: -0.48208 - size: (3, 15) - species 1 - id 1795\n",
      "Average adjusted fitness: 0.965\n",
      "Mean genetic distance 2.932, standard deviation 0.496\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 2.098 sec (2.829 average)\n",
      "\n",
      " ****** Running generation 8 ****** \n",
      "\n",
      "Population's average fitness: -40.74219 stdev: 122.59381\n",
      "Best fitness: -0.22572 - size: (4, 24) - species 1 - id 2104\n",
      "Average adjusted fitness: 0.954\n",
      "Mean genetic distance 2.939, standard deviation 0.553\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.873 sec (2.723 average)\n",
      "\n",
      " ****** Running generation 9 ****** \n",
      "\n",
      "Population's average fitness: -16.17513 stdev: 59.84621\n",
      "Best fitness: -0.17929 - size: (5, 25) - species 1 - id 2269\n",
      "Average adjusted fitness: 0.979\n",
      "Mean genetic distance 2.823, standard deviation 0.584\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.521 sec (2.603 average)\n",
      "Saving checkpoint to neat-checkpoint-9\n",
      "\n",
      " ****** Running generation 10 ****** \n",
      "\n",
      "Population's average fitness: -31.25026 stdev: 157.04944\n",
      "Best fitness: -0.17929 - size: (5, 25) - species 1 - id 2269\n",
      "Average adjusted fitness: 0.982\n",
      "Mean genetic distance 2.735, standard deviation 0.499\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.536 sec (2.404 average)\n",
      "\n",
      " ****** Running generation 11 ****** \n",
      "\n",
      "Population's average fitness: -22.96688 stdev: 170.09150\n",
      "Best fitness: -0.09482 - size: (4, 16) - species 1 - id 2883\n",
      "Average adjusted fitness: 0.989\n",
      "Mean genetic distance 2.523, standard deviation 0.514\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.150 sec (2.207 average)\n",
      "\n",
      " ****** Running generation 12 ****** \n",
      "\n",
      "Population's average fitness: -28.37486 stdev: 181.89495\n",
      "Best fitness: -0.09344 - size: (3, 14) - species 1 - id 3217\n",
      "Average adjusted fitness: 0.988\n",
      "Mean genetic distance 2.573, standard deviation 0.486\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.200 sec (2.036 average)\n",
      "\n",
      " ****** Running generation 13 ****** \n",
      "\n",
      "Population's average fitness: -58.32081 stdev: 314.62006\n",
      "Best fitness: -0.06424 - size: (4, 21) - species 1 - id 3418\n",
      "Average adjusted fitness: 0.984\n",
      "Mean genetic distance 2.572, standard deviation 0.426\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.199 sec (1.829 average)\n",
      "\n",
      " ****** Running generation 14 ****** \n",
      "\n",
      "Population's average fitness: -71.19706 stdev: 364.95310\n",
      "Best fitness: -0.06424 - size: (4, 21) - species 1 - id 3418\n",
      "Average adjusted fitness: 0.981\n",
      "Mean genetic distance 2.645, standard deviation 0.459\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.150 sec (1.672 average)\n",
      "Saving checkpoint to neat-checkpoint-14\n",
      "\n",
      " ****** Running generation 15 ****** \n",
      "\n",
      "Population's average fitness: -74.53174 stdev: 376.27607\n",
      "Best fitness: -0.06424 - size: (4, 21) - species 1 - id 3418\n",
      "Average adjusted fitness: 0.976\n",
      "Mean genetic distance 2.831, standard deviation 0.438\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.149 sec (1.518 average)\n",
      "\n",
      " ****** Running generation 16 ****** \n",
      "\n",
      "Population's average fitness: -43.28561 stdev: 323.37884\n",
      "Best fitness: -0.06424 - size: (4, 21) - species 1 - id 3418\n",
      "Average adjusted fitness: 0.986\n",
      "Mean genetic distance 2.756, standard deviation 0.444\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.250 sec (1.413 average)\n",
      "\n",
      " ****** Running generation 17 ****** \n",
      "\n",
      "Population's average fitness: -20.61827 stdev: 110.82215\n",
      "Best fitness: -0.06424 - size: (4, 21) - species 1 - id 3418\n",
      "Average adjusted fitness: 0.986\n",
      "Mean genetic distance 2.847, standard deviation 0.448\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.149 sec (1.318 average)\n",
      "\n",
      " ****** Running generation 18 ****** \n",
      "\n",
      "Population's average fitness: -45.04093 stdev: 370.90975\n",
      "Best fitness: -0.06424 - size: (4, 21) - species 1 - id 3418\n",
      "Average adjusted fitness: 0.992\n",
      "Mean genetic distance 2.885, standard deviation 0.389\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.065 sec (1.237 average)\n",
      "\n",
      " ****** Running generation 19 ****** \n",
      "\n",
      "Population's average fitness: -11.48486 stdev: 60.46796\n",
      "Best fitness: -0.06424 - size: (4, 21) - species 1 - id 3418\n",
      "Average adjusted fitness: 0.982\n",
      "Mean genetic distance 2.860, standard deviation 0.390\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.035 sec (1.188 average)\n",
      "Saving checkpoint to neat-checkpoint-19\n",
      "\n",
      " ****** Running generation 20 ****** \n",
      "\n",
      "Population's average fitness: -7.05834 stdev: 46.18688\n",
      "Best fitness: -0.06424 - size: (4, 21) - species 1 - id 3418\n",
      "Average adjusted fitness: 0.990\n",
      "Mean genetic distance 2.917, standard deviation 0.398\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.050 sec (1.140 average)\n",
      "\n",
      " ****** Running generation 21 ****** \n",
      "\n",
      "Population's average fitness: -38.00792 stdev: 328.47564\n",
      "Best fitness: -0.06424 - size: (4, 21) - species 1 - id 3418\n",
      "Average adjusted fitness: 0.991\n",
      "Mean genetic distance 2.904, standard deviation 0.439\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.237 sec (1.149 average)\n",
      "\n",
      " ****** Running generation 22 ****** \n",
      "\n",
      "Population's average fitness: -11.37794 stdev: 87.72003\n",
      "Best fitness: -0.06424 - size: (4, 21) - species 1 - id 3418\n",
      "Average adjusted fitness: 0.991\n",
      "Mean genetic distance 2.911, standard deviation 0.451\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.079 sec (1.136 average)\n",
      "\n",
      " ****** Running generation 23 ****** \n",
      "\n",
      "Population's average fitness: -12.66888 stdev: 107.71507\n",
      "Best fitness: -0.06424 - size: (4, 21) - species 1 - id 3418\n",
      "Average adjusted fitness: 0.992\n",
      "Mean genetic distance 2.877, standard deviation 0.524\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.248 sec (1.141 average)\n",
      "\n",
      " ****** Running generation 24 ****** \n",
      "\n",
      "Population's average fitness: -11.30004 stdev: 111.17477\n",
      "Best fitness: -0.06423 - size: (4, 18) - species 1 - id 6066\n",
      "Average adjusted fitness: 0.994\n",
      "Mean genetic distance 2.831, standard deviation 0.428\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.119 sec (1.138 average)\n",
      "Saving checkpoint to neat-checkpoint-24\n",
      "\n",
      " ****** Running generation 25 ****** \n",
      "\n",
      "Population's average fitness: -13.12232 stdev: 79.38757\n",
      "Best fitness: -0.06423 - size: (4, 18) - species 1 - id 6066\n",
      "Average adjusted fitness: 0.988\n",
      "Mean genetic distance 2.831, standard deviation 0.422\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.368 sec (1.160 average)\n",
      "\n",
      " ****** Running generation 26 ****** \n",
      "\n",
      "Population's average fitness: -15.35309 stdev: 112.66292\n",
      "Best fitness: -0.06423 - size: (4, 18) - species 1 - id 6066\n",
      "Average adjusted fitness: 0.991\n",
      "Mean genetic distance 2.832, standard deviation 0.427\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.334 sec (1.168 average)\n",
      "\n",
      " ****** Running generation 27 ****** \n",
      "\n",
      "Population's average fitness: -11.88008 stdev: 72.66696\n",
      "Best fitness: -0.06423 - size: (4, 18) - species 1 - id 6066\n",
      "Average adjusted fitness: 0.989\n",
      "Mean genetic distance 2.929, standard deviation 0.425\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.367 sec (1.190 average)\n",
      "\n",
      " ****** Running generation 28 ****** \n",
      "\n",
      "Population's average fitness: -10.58712 stdev: 50.59950\n",
      "Best fitness: -0.06423 - size: (4, 18) - species 1 - id 6066\n",
      "Average adjusted fitness: 0.984\n",
      "Mean genetic distance 2.739, standard deviation 0.405\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.248 sec (1.208 average)\n",
      "\n",
      " ****** Running generation 29 ****** \n",
      "\n",
      "Population's average fitness: -31.62477 stdev: 152.63977\n",
      "Best fitness: -0.06423 - size: (4, 18) - species 1 - id 6066\n",
      "Average adjusted fitness: 0.981\n",
      "Mean genetic distance 2.684, standard deviation 0.300\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.335 sec (1.238 average)\n",
      "Saving checkpoint to neat-checkpoint-29\n",
      "\n",
      " ****** Running generation 30 ****** \n",
      "\n",
      "Population's average fitness: -25.07301 stdev: 171.23110\n",
      "Best fitness: -0.06423 - size: (4, 18) - species 1 - id 6066\n",
      "Average adjusted fitness: 0.989\n",
      "Mean genetic distance 2.774, standard deviation 0.339\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.352 sec (1.269 average)\n",
      "\n",
      " ****** Running generation 31 ****** \n",
      "\n",
      "Population's average fitness: -38.47705 stdev: 255.81678\n",
      "Best fitness: -0.06423 - size: (4, 18) - species 1 - id 6066\n",
      "Average adjusted fitness: 0.987\n",
      "Mean genetic distance 2.769, standard deviation 0.355\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.683 sec (1.313 average)\n",
      "\n",
      " ****** Running generation 32 ****** \n",
      "\n",
      "Population's average fitness: -38.81501 stdev: 428.95366\n",
      "Best fitness: -0.06423 - size: (4, 18) - species 1 - id 6066\n",
      "Average adjusted fitness: 0.994\n",
      "Mean genetic distance 2.677, standard deviation 0.375\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.517 sec (1.357 average)\n",
      "\n",
      " ****** Running generation 33 ****** \n",
      "\n",
      "Population's average fitness: -38.37701 stdev: 297.81126\n",
      "Best fitness: -0.06423 - size: (4, 18) - species 1 - id 6066\n",
      "Average adjusted fitness: 0.991\n",
      "Mean genetic distance 2.669, standard deviation 0.366\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.616 sec (1.394 average)\n",
      "\n",
      " ****** Running generation 34 ****** \n",
      "\n",
      "Population's average fitness: -11.21363 stdev: 48.76591\n",
      "Best fitness: -0.06423 - size: (4, 18) - species 1 - id 6066\n",
      "Average adjusted fitness: 0.975\n",
      "Mean genetic distance 2.631, standard deviation 0.373\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.431 sec (1.425 average)\n",
      "Saving checkpoint to neat-checkpoint-34\n",
      "\n",
      " ****** Running generation 35 ****** \n",
      "\n",
      "Population's average fitness: -26.32014 stdev: 144.22757\n",
      "Best fitness: -0.06423 - size: (4, 18) - species 1 - id 6066\n",
      "Average adjusted fitness: 0.978\n",
      "Mean genetic distance 2.647, standard deviation 0.367\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.481 sec (1.436 average)\n",
      "\n",
      " ****** Running generation 36 ****** \n",
      "\n",
      "Population's average fitness: -70.38636 stdev: 350.70623\n",
      "Best fitness: -0.06423 - size: (4, 18) - species 1 - id 6066\n",
      "Average adjusted fitness: 0.980\n",
      "Mean genetic distance 2.627, standard deviation 0.391\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.459 sec (1.449 average)\n",
      "\n",
      " ****** Running generation 37 ****** \n",
      "\n",
      "Population's average fitness: -33.57143 stdev: 144.41538\n",
      "Best fitness: -0.06423 - size: (4, 18) - species 1 - id 6066\n",
      "Average adjusted fitness: 0.976\n",
      "Mean genetic distance 2.662, standard deviation 0.307\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.381 sec (1.450 average)\n",
      "\n",
      " ****** Running generation 38 ****** \n",
      "\n",
      "Population's average fitness: -24.76182 stdev: 175.24071\n",
      "Best fitness: -0.06423 - size: (4, 18) - species 1 - id 6066\n",
      "Average adjusted fitness: 0.990\n",
      "Mean genetic distance 2.643, standard deviation 0.366\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.374 sec (1.463 average)\n",
      "\n",
      " ****** Running generation 39 ****** \n",
      "\n",
      "Population's average fitness: -46.94837 stdev: 336.49901\n",
      "Best fitness: -0.06423 - size: (4, 18) - species 1 - id 6066\n",
      "Average adjusted fitness: 0.989\n",
      "Mean genetic distance 2.683, standard deviation 0.354\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.396 sec (1.469 average)\n",
      "Saving checkpoint to neat-checkpoint-39\n",
      "\n",
      " ****** Running generation 40 ****** \n",
      "\n",
      "Population's average fitness: -17.13047 stdev: 82.13279\n",
      "Best fitness: -0.06400 - size: (11, 17) - species 1 - id 10148\n",
      "Average adjusted fitness: 0.982\n",
      "Mean genetic distance 2.806, standard deviation 0.254\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.535 sec (1.487 average)\n",
      "\n",
      " ****** Running generation 41 ****** \n",
      "\n",
      "Population's average fitness: -34.20659 stdev: 200.99755\n",
      "Best fitness: -0.06400 - size: (11, 17) - species 1 - id 10148\n",
      "Average adjusted fitness: 0.988\n",
      "Mean genetic distance 2.952, standard deviation 0.283\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.378 sec (1.457 average)\n",
      "\n",
      " ****** Running generation 42 ****** \n",
      "\n",
      "Population's average fitness: -36.19759 stdev: 236.39578\n",
      "Best fitness: -0.06400 - size: (11, 17) - species 1 - id 10148\n",
      "Average adjusted fitness: 0.987\n",
      "Mean genetic distance 3.152, standard deviation 0.330\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.872 sec (1.492 average)\n",
      "\n",
      " ****** Running generation 43 ****** \n",
      "\n",
      "Population's average fitness: -36.79631 stdev: 231.45631\n",
      "Best fitness: -0.06400 - size: (11, 17) - species 1 - id 10148\n",
      "Average adjusted fitness: 0.985\n",
      "Mean genetic distance 3.219, standard deviation 0.339\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.567 sec (1.487 average)\n",
      "\n",
      " ****** Running generation 44 ****** \n",
      "\n",
      "Population's average fitness: -50.02418 stdev: 443.65750\n",
      "Best fitness: -0.06400 - size: (11, 17) - species 1 - id 10148\n",
      "Average adjusted fitness: 0.993\n",
      "Mean genetic distance 2.881, standard deviation 0.515\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.517 sec (1.496 average)\n",
      "Saving checkpoint to neat-checkpoint-44\n",
      "\n",
      " ****** Running generation 45 ****** \n",
      "\n",
      "Population's average fitness: -68.51976 stdev: 383.86943\n",
      "Best fitness: -0.06400 - size: (11, 17) - species 1 - id 10148\n",
      "Average adjusted fitness: 0.983\n",
      "Mean genetic distance 2.739, standard deviation 0.435\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.866 sec (1.534 average)\n",
      "\n",
      " ****** Running generation 46 ****** \n",
      "\n",
      "Population's average fitness: -37.49097 stdev: 223.55201\n",
      "Best fitness: -0.06400 - size: (11, 17) - species 1 - id 10148\n",
      "Average adjusted fitness: 0.987\n",
      "Mean genetic distance 2.782, standard deviation 0.418\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.600 sec (1.549 average)\n",
      "\n",
      " ****** Running generation 47 ****** \n",
      "\n",
      "Population's average fitness: -48.07631 stdev: 405.60567\n",
      "Best fitness: -0.06400 - size: (11, 17) - species 1 - id 10148\n",
      "Average adjusted fitness: 0.992\n",
      "Mean genetic distance 2.786, standard deviation 0.439\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.461 sec (1.557 average)\n",
      "\n",
      " ****** Running generation 48 ****** \n",
      "\n",
      "Population's average fitness: -28.79978 stdev: 161.68935\n",
      "Best fitness: -0.06400 - size: (11, 17) - species 1 - id 10148\n",
      "Average adjusted fitness: 0.985\n",
      "Mean genetic distance 2.835, standard deviation 0.330\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.440 sec (1.563 average)\n",
      "\n",
      " ****** Running generation 49 ****** \n",
      "\n",
      "Population's average fitness: -40.54254 stdev: 161.88754\n",
      "Best fitness: -0.06400 - size: (11, 17) - species 1 - id 10148\n",
      "Average adjusted fitness: 0.972\n",
      "Mean genetic distance 2.871, standard deviation 0.363\n",
      "Population of 250 members in 1 species\n",
      "Total extinctions: 0\n",
      "Generation time: 1.616 sec (1.585 average)\n",
      "Saving checkpoint to neat-checkpoint-49\n"
     ]
    }
   ],
   "source": [
    "winner = pp.run(eval_genomes, 50)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **5. Results**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **5.1. Get a visual interpretation.**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Dictionary for the features:** <br>\n",
    "lactation: 'lct' <br>\n",
    "delivery_age_years: 'dlv_ay' <br>\n",
    "'dim' <br>\n",
    "dry_days: 'dd' <br>\n",
    "forage_kg_day: 'fkd' <br>\n",
    "rumination_min_day: 'rmd' <br>\n",
    "milk_kg_day: 'mkd' <br>\n",
    "milk_kg_min_robot: 'mkmr' <br>\n",
    "milkings_day: 'md' <br>\n",
    "errors_by_100_milkings: 'er_100m' <br>\n",
    "high_cdt_by_100_milkings: 'hcdt_100m' <br>\n",
    "watery_by_100_milkings: 'w100m' <br>\n",
    "refusals_by_milking: 'ref_bm' <br>\n",
    "colostrum_separated_kg: 'cskg' <br>\n",
    "\n",
    "'lct', 'dlv_ay', 'dim', 'dd', 'fkd', 'rmd', 'mkd', 'mkmr', 'md', 'er_100m', 'hcdt_100m', 'w100m', ref_bm', 'cskg'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 8.1.0 (20230707.0739)\n",
       " -->\n",
       "<!-- Pages: 1 -->\n",
       "<svg width=\"708pt\" height=\"373pt\"\n",
       " viewBox=\"0.00 0.00 708.25 373.13\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(1 1) rotate(0) translate(4 369.13)\">\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-369.13 704.25,-369.13 704.25,4 -4,4\"/>\n",
       "<!-- lct -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>lct</title>\n",
       "<polygon fill=\"lightgray\" stroke=\"black\" points=\"25,-365.13 0,-365.13 0,-345.88 25,-345.88 25,-365.13\"/>\n",
       "<text text-anchor=\"middle\" x=\"12.5\" y=\"-352.58\" font-family=\"Times New Roman,serif\" font-size=\"9.00\">lct</text>\n",
       "</g>\n",
       "<!-- dlv_ay -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>dlv_ay</title>\n",
       "<polygon fill=\"lightgray\" stroke=\"black\" points=\"83.88,-365.13 43.12,-365.13 43.12,-345.88 83.88,-345.88 83.88,-365.13\"/>\n",
       "<text text-anchor=\"middle\" x=\"63.5\" y=\"-352.58\" font-family=\"Times New Roman,serif\" font-size=\"9.00\">dlv_ay</text>\n",
       "</g>\n",
       "<!-- dim -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>dim</title>\n",
       "<polygon fill=\"lightgray\" stroke=\"black\" points=\"131.25,-365.13 101.75,-365.13 101.75,-345.88 131.25,-345.88 131.25,-365.13\"/>\n",
       "<text text-anchor=\"middle\" x=\"116.5\" y=\"-352.58\" font-family=\"Times New Roman,serif\" font-size=\"9.00\">dim</text>\n",
       "</g>\n",
       "<!-- 4893 -->\n",
       "<g id=\"node20\" class=\"node\">\n",
       "<title>4893</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"139.5\" cy=\"-285.84\" rx=\"24.04\" ry=\"24.04\"/>\n",
       "<text text-anchor=\"middle\" x=\"139.5\" y=\"-282.91\" font-family=\"Times New Roman,serif\" font-size=\"9.00\">4893</text>\n",
       "</g>\n",
       "<!-- dim&#45;&gt;4893 -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>dim&#45;&gt;4893</title>\n",
       "<path fill=\"none\" stroke=\"red\" stroke-width=\"0.2\" d=\"M119.51,-345.66C121.89,-338.65 125.37,-328.42 128.76,-318.45\"/>\n",
       "<polygon fill=\"red\" stroke=\"red\" stroke-width=\"0.2\" points=\"132.04,-319.67 131.94,-309.08 125.41,-317.42 132.04,-319.67\"/>\n",
       "</g>\n",
       "<!-- dd -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>dd</title>\n",
       "<polygon fill=\"lightgray\" stroke=\"black\" points=\"219,-211.38 194,-211.38 194,-192.13 219,-192.13 219,-211.38\"/>\n",
       "<text text-anchor=\"middle\" x=\"206.5\" y=\"-198.83\" font-family=\"Times New Roman,serif\" font-size=\"9.00\">dd</text>\n",
       "</g>\n",
       "<!-- 342 -->\n",
       "<g id=\"node19\" class=\"node\">\n",
       "<title>342</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"186.5\" cy=\"-117.67\" rx=\"20.86\" ry=\"20.86\"/>\n",
       "<text text-anchor=\"middle\" x=\"186.5\" y=\"-114.75\" font-family=\"Times New Roman,serif\" font-size=\"9.00\">342</text>\n",
       "</g>\n",
       "<!-- dd&#45;&gt;342 -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>dd&#45;&gt;342</title>\n",
       "<path fill=\"none\" stroke=\"green\" stroke-width=\"0.32\" d=\"M204.28,-191.66C201.73,-181.18 197.41,-163.43 193.66,-148.06\"/>\n",
       "<polygon fill=\"green\" stroke=\"green\" stroke-width=\"0.32\" points=\"197.08,-147.29 191.31,-138.41 190.28,-148.95 197.08,-147.29\"/>\n",
       "</g>\n",
       "<!-- fkd -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>fkd</title>\n",
       "<polygon fill=\"lightgray\" stroke=\"black\" points=\"352.5,-365.13 324.5,-365.13 324.5,-345.88 352.5,-345.88 352.5,-365.13\"/>\n",
       "<text text-anchor=\"middle\" x=\"338.5\" y=\"-352.58\" font-family=\"Times New Roman,serif\" font-size=\"9.00\">fkd</text>\n",
       "</g>\n",
       "<!-- 2629 -->\n",
       "<g id=\"node15\" class=\"node\">\n",
       "<title>2629</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"338.5\" cy=\"-117.67\" rx=\"24.04\" ry=\"24.04\"/>\n",
       "<text text-anchor=\"middle\" x=\"338.5\" y=\"-114.75\" font-family=\"Times New Roman,serif\" font-size=\"9.00\">2629</text>\n",
       "</g>\n",
       "<!-- fkd&#45;&gt;2629 -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>fkd&#45;&gt;2629</title>\n",
       "<path fill=\"none\" stroke=\"green\" stroke-width=\"0.26\" d=\"M336.7,-345.66C334.97,-336.69 332.48,-322.39 331.5,-309.88 326.93,-251.32 328.04,-236.35 331.5,-177.71 332,-169.29 332.89,-160.22 333.88,-151.78\"/>\n",
       "<polygon fill=\"green\" stroke=\"green\" stroke-width=\"0.26\" points=\"337.35,-152.24 335.11,-141.89 330.4,-151.38 337.35,-152.24\"/>\n",
       "</g>\n",
       "<!-- 6197 -->\n",
       "<g id=\"node18\" class=\"node\">\n",
       "<title>6197</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"364.5\" cy=\"-285.84\" rx=\"24.04\" ry=\"24.04\"/>\n",
       "<text text-anchor=\"middle\" x=\"364.5\" y=\"-282.91\" font-family=\"Times New Roman,serif\" font-size=\"9.00\">6197</text>\n",
       "</g>\n",
       "<!-- fkd&#45;&gt;6197 -->\n",
       "<g id=\"edge11\" class=\"edge\">\n",
       "<title>fkd&#45;&gt;6197</title>\n",
       "<path fill=\"none\" stroke=\"green\" stroke-width=\"0.32\" d=\"M341.9,-345.66C344.64,-338.53 348.66,-328.05 352.56,-317.91\"/>\n",
       "<polygon fill=\"green\" stroke=\"green\" stroke-width=\"0.32\" points=\"355.75,-319.38 356.07,-308.79 349.21,-316.87 355.75,-319.38\"/>\n",
       "</g>\n",
       "<!-- rmd -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>rmd</title>\n",
       "<polygon fill=\"lightgray\" stroke=\"black\" points=\"400.62,-365.13 370.38,-365.13 370.38,-345.88 400.62,-345.88 400.62,-365.13\"/>\n",
       "<text text-anchor=\"middle\" x=\"385.5\" y=\"-352.58\" font-family=\"Times New Roman,serif\" font-size=\"9.00\">rmd</text>\n",
       "</g>\n",
       "<!-- mkd -->\n",
       "<g id=\"node7\" class=\"node\">\n",
       "<title>mkd</title>\n",
       "<polygon fill=\"lightgray\" stroke=\"black\" points=\"450.38,-365.13 418.62,-365.13 418.62,-345.88 450.38,-345.88 450.38,-365.13\"/>\n",
       "<text text-anchor=\"middle\" x=\"434.5\" y=\"-352.58\" font-family=\"Times New Roman,serif\" font-size=\"9.00\">mkd</text>\n",
       "</g>\n",
       "<!-- mkmr -->\n",
       "<g id=\"node8\" class=\"node\">\n",
       "<title>mkmr</title>\n",
       "<polygon fill=\"lightgray\" stroke=\"black\" points=\"505,-365.13 468,-365.13 468,-345.88 505,-345.88 505,-365.13\"/>\n",
       "<text text-anchor=\"middle\" x=\"486.5\" y=\"-352.58\" font-family=\"Times New Roman,serif\" font-size=\"9.00\">mkmr</text>\n",
       "</g>\n",
       "<!-- md -->\n",
       "<g id=\"node9\" class=\"node\">\n",
       "<title>md</title>\n",
       "<polygon fill=\"lightgray\" stroke=\"black\" points=\"176.12,-365.13 148.88,-365.13 148.88,-345.88 176.12,-345.88 176.12,-365.13\"/>\n",
       "<text text-anchor=\"middle\" x=\"162.5\" y=\"-352.58\" font-family=\"Times New Roman,serif\" font-size=\"9.00\">md</text>\n",
       "</g>\n",
       "<!-- md&#45;&gt;4893 -->\n",
       "<g id=\"edge8\" class=\"edge\">\n",
       "<title>md&#45;&gt;4893</title>\n",
       "<path fill=\"none\" stroke=\"green\" stroke-width=\"0.25\" d=\"M159.49,-345.66C157.11,-338.65 153.63,-328.42 150.24,-318.45\"/>\n",
       "<polygon fill=\"green\" stroke=\"green\" stroke-width=\"0.25\" points=\"153.59,-317.42 147.06,-309.08 146.96,-319.67 153.59,-317.42\"/>\n",
       "</g>\n",
       "<!-- er_100m -->\n",
       "<g id=\"node10\" class=\"node\">\n",
       "<title>er_100m</title>\n",
       "<polygon fill=\"lightgray\" stroke=\"black\" points=\"284.25,-211.38 236.75,-211.38 236.75,-192.13 284.25,-192.13 284.25,-211.38\"/>\n",
       "<text text-anchor=\"middle\" x=\"260.5\" y=\"-198.83\" font-family=\"Times New Roman,serif\" font-size=\"9.00\">er_100m</text>\n",
       "</g>\n",
       "<!-- er_100m&#45;&gt;342 -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>er_100m&#45;&gt;342</title>\n",
       "<path fill=\"none\" stroke=\"red\" stroke-width=\"0.18\" d=\"M252.3,-191.66C241.4,-179.57 221.77,-157.8 206.78,-141.17\"/>\n",
       "<polygon fill=\"red\" stroke=\"red\" stroke-width=\"0.18\" points=\"209.57,-139.03 200.27,-133.95 204.37,-143.72 209.57,-139.03\"/>\n",
       "</g>\n",
       "<!-- hcdt_100m -->\n",
       "<g id=\"node11\" class=\"node\">\n",
       "<title>hcdt_100m</title>\n",
       "<polygon fill=\"lightgray\" stroke=\"black\" points=\"579.75,-365.13 523.25,-365.13 523.25,-345.88 579.75,-345.88 579.75,-365.13\"/>\n",
       "<text text-anchor=\"middle\" x=\"551.5\" y=\"-352.58\" font-family=\"Times New Roman,serif\" font-size=\"9.00\">hcdt_100m</text>\n",
       "</g>\n",
       "<!-- w100m -->\n",
       "<g id=\"node12\" class=\"node\">\n",
       "<title>w100m</title>\n",
       "<polygon fill=\"lightgray\" stroke=\"black\" points=\"641,-365.13 598,-365.13 598,-345.88 641,-345.88 641,-365.13\"/>\n",
       "<text text-anchor=\"middle\" x=\"619.5\" y=\"-352.58\" font-family=\"Times New Roman,serif\" font-size=\"9.00\">w100m</text>\n",
       "</g>\n",
       "<!-- ref_bm -->\n",
       "<g id=\"node13\" class=\"node\">\n",
       "<title>ref_bm</title>\n",
       "<polygon fill=\"lightgray\" stroke=\"black\" points=\"700.25,-365.13 658.75,-365.13 658.75,-345.88 700.25,-345.88 700.25,-365.13\"/>\n",
       "<text text-anchor=\"middle\" x=\"679.5\" y=\"-352.58\" font-family=\"Times New Roman,serif\" font-size=\"9.00\">ref_bm</text>\n",
       "</g>\n",
       "<!-- NEAT -->\n",
       "<g id=\"node14\" class=\"node\">\n",
       "<title>NEAT</title>\n",
       "<ellipse fill=\"lightblue\" stroke=\"black\" cx=\"186.5\" cy=\"-28.81\" rx=\"28.81\" ry=\"28.81\"/>\n",
       "<text text-anchor=\"middle\" x=\"186.5\" y=\"-25.89\" font-family=\"Times New Roman,serif\" font-size=\"9.00\">NEAT</text>\n",
       "</g>\n",
       "<!-- 5863 -->\n",
       "<g id=\"node16\" class=\"node\">\n",
       "<title>5863</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"113.5\" cy=\"-201.75\" rx=\"24.04\" ry=\"24.04\"/>\n",
       "<text text-anchor=\"middle\" x=\"113.5\" y=\"-198.83\" font-family=\"Times New Roman,serif\" font-size=\"9.00\">5863</text>\n",
       "</g>\n",
       "<!-- 5863&#45;&gt;342 -->\n",
       "<g id=\"edge10\" class=\"edge\">\n",
       "<title>5863&#45;&gt;342</title>\n",
       "<path fill=\"none\" stroke=\"red\" stroke-width=\"1.03\" d=\"M128.98,-183.35C139.69,-171.3 154.07,-155.14 165.77,-141.98\"/>\n",
       "<polygon fill=\"red\" stroke=\"red\" stroke-width=\"1.03\" points=\"167.97,-144.64 172,-134.85 162.74,-139.99 167.97,-144.64\"/>\n",
       "</g>\n",
       "<!-- 5136 -->\n",
       "<g id=\"node17\" class=\"node\">\n",
       "<title>5136</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"364.5\" cy=\"-201.75\" rx=\"24.04\" ry=\"24.04\"/>\n",
       "<text text-anchor=\"middle\" x=\"364.5\" y=\"-198.83\" font-family=\"Times New Roman,serif\" font-size=\"9.00\">5136</text>\n",
       "</g>\n",
       "<!-- 5136&#45;&gt;2629 -->\n",
       "<g id=\"edge7\" class=\"edge\">\n",
       "<title>5136&#45;&gt;2629</title>\n",
       "<path fill=\"none\" stroke=\"red\" stroke-width=\"0.44\" d=\"M357.4,-178.32C354.67,-169.73 351.51,-159.75 348.56,-150.44\"/>\n",
       "<polygon fill=\"red\" stroke=\"red\" stroke-width=\"0.44\" points=\"351.95,-149.55 345.6,-141.07 345.28,-151.66 351.95,-149.55\"/>\n",
       "</g>\n",
       "<!-- 6197&#45;&gt;5136 -->\n",
       "<g id=\"edge12\" class=\"edge\">\n",
       "<title>6197&#45;&gt;5136</title>\n",
       "<path fill=\"none\" stroke=\"green\" stroke-width=\"0.26\" d=\"M364.5,-261.54C364.5,-253.56 364.5,-244.49 364.5,-235.89\"/>\n",
       "<polygon fill=\"green\" stroke=\"green\" stroke-width=\"0.26\" points=\"368,-236.11 364.5,-226.11 361,-236.11 368,-236.11\"/>\n",
       "</g>\n",
       "<!-- 342&#45;&gt;NEAT -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>342&#45;&gt;NEAT</title>\n",
       "<path fill=\"none\" stroke=\"green\" stroke-width=\"0.43\" d=\"M186.5,-96.53C186.5,-87.96 186.5,-77.67 186.5,-67.77\"/>\n",
       "<polygon fill=\"green\" stroke=\"green\" stroke-width=\"0.43\" points=\"190,-68.07 186.5,-58.07 183,-68.07 190,-68.07\"/>\n",
       "</g>\n",
       "<!-- 4893&#45;&gt;5863 -->\n",
       "<g id=\"edge9\" class=\"edge\">\n",
       "<title>4893&#45;&gt;5863</title>\n",
       "<path fill=\"none\" stroke=\"green\" stroke-width=\"0.18\" d=\"M132.4,-262.41C129.67,-253.81 126.51,-243.83 123.56,-234.53\"/>\n",
       "<polygon fill=\"green\" stroke=\"green\" stroke-width=\"0.18\" points=\"126.95,-233.63 120.6,-225.16 120.28,-235.75 126.95,-233.63\"/>\n",
       "</g>\n",
       "<!-- 4893&#45;&gt;342 -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>4893&#45;&gt;342</title>\n",
       "<path fill=\"none\" stroke=\"red\" stroke-width=\"0.88\" d=\"M145.84,-262.43C154.17,-232.98 168.82,-181.17 178.08,-148.45\"/>\n",
       "<polygon fill=\"red\" stroke=\"red\" stroke-width=\"0.88\" points=\"181.64,-149.72 180.99,-139.15 174.9,-147.82 181.64,-149.72\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x25bcac34c70>"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_fts = ['lct', 'dlv_ay', 'dim', 'dd', 'fkd', 'rmd', 'mkd', 'mkmr', 'md', 'er_100m', 'hcdt_100m', 'w100m', 'ref_bm', 'cskg']\n",
    "input_names = {0: 'NEAT'}\n",
    "for i in range(len(data_fts)):\n",
    "    input_names.update({-1*(i+1): data_fts[i]})\n",
    "\n",
    "draw_net(config, winner, True, node_names=input_names, prune_unused=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **5.2. Check *best/winning genome***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Best genome:\n",
      "Key: 10148\n",
      "Fitness: -0.06399571159637707\n",
      "Nodes:\n",
      "\t0 DefaultNodeGene(key=0, bias=4.88223828327583, response=1.0, activation=relu, aggregation=mean)\n",
      "\t342 DefaultNodeGene(key=342, bias=1.4931925098480672, response=1.0, activation=gauss, aggregation=sum)\n",
      "\t2629 DefaultNodeGene(key=2629, bias=1.1762333733238584, response=1.0, activation=relu, aggregation=mean)\n",
      "\t3331 DefaultNodeGene(key=3331, bias=0.06331043652925228, response=1.0, activation=identity, aggregation=mean)\n",
      "\t3354 DefaultNodeGene(key=3354, bias=0.17559195579302161, response=1.0, activation=gauss, aggregation=sum)\n",
      "\t4893 DefaultNodeGene(key=4893, bias=-1.4178321445674134, response=1.0, activation=relu, aggregation=sum)\n",
      "\t5003 DefaultNodeGene(key=5003, bias=2.14509318438079, response=1.0, activation=identity, aggregation=mean)\n",
      "\t5136 DefaultNodeGene(key=5136, bias=-0.4298815238811672, response=1.0, activation=relu, aggregation=sum)\n",
      "\t5344 DefaultNodeGene(key=5344, bias=1.7850195963679687, response=1.0, activation=gauss, aggregation=sum)\n",
      "\t5863 DefaultNodeGene(key=5863, bias=0.5813071594125879, response=1.0, activation=relu, aggregation=sum)\n",
      "\t6197 DefaultNodeGene(key=6197, bias=0.012035842874213803, response=1.0, activation=identity, aggregation=sum)\n",
      "Connections:\n",
      "\tDefaultConnectionGene(key=(-13, 342), weight=5.2709817448251295, enabled=False)\n",
      "\tDefaultConnectionGene(key=(-10, 342), weight=-0.38771512285148546, enabled=True)\n",
      "\tDefaultConnectionGene(key=(-9, 4893), weight=0.726657950073001, enabled=True)\n",
      "\tDefaultConnectionGene(key=(-8, 342), weight=0.4762728766558091, enabled=False)\n",
      "\tDefaultConnectionGene(key=(-7, 342), weight=0.6822700090751442, enabled=False)\n",
      "\tDefaultConnectionGene(key=(-5, 342), weight=-1.070197368076427, enabled=False)\n",
      "\tDefaultConnectionGene(key=(-5, 2629), weight=0.8046560881344926, enabled=True)\n",
      "\tDefaultConnectionGene(key=(-5, 3331), weight=1.7894148362250033, enabled=True)\n",
      "\tDefaultConnectionGene(key=(-5, 6197), weight=1.1128431200849649, enabled=True)\n",
      "\tDefaultConnectionGene(key=(-4, 342), weight=1.1125084001090246, enabled=True)\n",
      "\tDefaultConnectionGene(key=(-3, 342), weight=0.5009283863695905, enabled=False)\n",
      "\tDefaultConnectionGene(key=(-3, 4893), weight=-0.5216568316507917, enabled=True)\n",
      "\tDefaultConnectionGene(key=(-3, 5003), weight=-1.3914742875498944, enabled=True)\n",
      "\tDefaultConnectionGene(key=(-2, 3331), weight=2.1465486341174254, enabled=False)\n",
      "\tDefaultConnectionGene(key=(-2, 3354), weight=-0.21248232479912071, enabled=True)\n",
      "\tDefaultConnectionGene(key=(-2, 5344), weight=-0.5745499577062511, enabled=True)\n",
      "\tDefaultConnectionGene(key=(-1, 342), weight=-1.0212434687032168, enabled=False)\n",
      "\tDefaultConnectionGene(key=(342, 0), weight=1.6379008540369555, enabled=True)\n",
      "\tDefaultConnectionGene(key=(2629, 342), weight=1.991143281689543, enabled=False)\n",
      "\tDefaultConnectionGene(key=(4893, 342), weight=-3.8957501673377894, enabled=True)\n",
      "\tDefaultConnectionGene(key=(4893, 5863), weight=0.3889069459893507, enabled=True)\n",
      "\tDefaultConnectionGene(key=(5136, 2629), weight=-1.6775393653074466, enabled=True)\n",
      "\tDefaultConnectionGene(key=(5344, 3331), weight=-0.6449058833434778, enabled=True)\n",
      "\tDefaultConnectionGene(key=(5863, 342), weight=-4.645134157953874, enabled=True)\n",
      "\tDefaultConnectionGene(key=(6197, 5136), weight=0.790024346798164, enabled=True)\n"
     ]
    }
   ],
   "source": [
    "print('\\nBest genome:\\n{!s}'.format(winner))\n",
    "\n",
    "# Needs to be converted to nn (on Outputs cell)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **5.3. Outputs**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Output:\n"
     ]
    }
   ],
   "source": [
    "# Show output of the most fit genome against training data.\n",
    "print('\\nOutput:')\n",
    "\n",
    "winner_net = neat.nn.FeedForwardNetwork.create(winner, config)\n",
    "\n",
    "for xi, xo in zip(X_train, y_train):\n",
    "# for xi, xo in zip(xor_inputs, xor_outputs):\n",
    "    output = winner_net.activate(xi)\n",
    "    #print(\"  expected output {!r}, got {!r}\".format(xo, output))\n",
    "    \n",
    "out_train = []\n",
    "for xi in X_train:\n",
    "    output = winner_net.activate(xi)\n",
    "    out_train.append(output[0])\n",
    "\n",
    "out_val = []\n",
    "for xi in X_val:\n",
    "    output = winner_net.activate(xi)\n",
    "    out_val.append(output[0])\n",
    "\n",
    "# TEST predictions\n",
    "out_test = []\n",
    "for xi in X_test:\n",
    "    output = winner_net.activate(xi)\n",
    "    out_test.append(output[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[4.88223828327583,\n",
       " 4.88223828327583,\n",
       " 4.88223828327583,\n",
       " 4.88223828327583,\n",
       " 4.882238283275832,\n",
       " 4.88223828327583,\n",
       " 4.882334998664991,\n",
       " 4.88223828327583,\n",
       " 4.88223828327583,\n",
       " 4.88223828327583,\n",
       " 4.88223828327583,\n",
       " 4.88223828327583,\n",
       " 4.88223828327583,\n",
       " 4.88223828327583,\n",
       " 4.88223828327583,\n",
       " 4.88223828327583,\n",
       " 4.88223828327583,\n",
       " 4.882238283275891,\n",
       " 4.88223828327583,\n",
       " 4.88223828327583,\n",
       " 4.88223828327583,\n",
       " 4.88223828327583,\n",
       " 4.88223828327583,\n",
       " 4.88223828344273,\n",
       " 4.88223828327583,\n",
       " 4.88223828327583,\n",
       " 4.88223828327583,\n",
       " 4.88223828327583,\n",
       " 4.89190811913396,\n",
       " 4.88223828327583,\n",
       " 4.88223828327583,\n",
       " 4.88223828327583,\n",
       " 4.88223828327583,\n",
       " 4.88223828327583,\n",
       " 4.88223828327583,\n",
       " 4.88223828327583,\n",
       " 4.88223828327583,\n",
       " 4.88223828327583,\n",
       " 4.88223828327583,\n",
       " 4.88223828327583,\n",
       " 4.88223828327583,\n",
       " 4.88223828327583,\n",
       " 4.88223828327583,\n",
       " 4.88223828327583,\n",
       " 4.88223828327583]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out_test"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **5.4. RMSE**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RMSE: 0.07405593163525723\n"
     ]
    }
   ],
   "source": [
    "expected_output = y_test \n",
    "got_output = out_test\n",
    "\n",
    "rmse = np.sqrt(mean_squared_error(expected_output, got_output))\n",
    "print(\"RMSE:\", rmse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.062495"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Re-run the notebook and changing the seed - to get different RMSE values\n",
    "# (seed=6, 8, 11, 12, 21, 27, 29, 33, 66, 91)\n",
    "# list of 10 RMSE-test values\n",
    "list_RMSE = [0.06421294813838227, 0.052080539691019494, 0.06241040499372979, 0.059267949283075196, 0.05692388766008969,\n",
    "             0.06820785744466334, 0.06257905782196159, 0.0605951983058181, 0.0745904275107406, 0.07405593163525723]\n",
    "\n",
    "# compute the median of list_RMSE\n",
    "median_RMSE = sts.median(list_RMSE)\n",
    "round(median_RMSE, 6)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> **Median RMSE on test set: 0.062495** <br>\n",
    ">\n",
    "> Achieved with 50 generations and the parameters on *config-feedforward-neat*.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Visualize"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjcAAAHFCAYAAAAOmtghAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAABPC0lEQVR4nO3deVxUVf8H8M9lBmbYUXYEBZfUcimxzC0xFXNLK9O03NLSMBPJ3cq1MJ9SslIrF1rNp1Izt6Ry/alPLpDm0uKGJYjgwiYwzJzfH+NMTqCyzNw7zHzer9e8YO6ce+c7B9SP55x7rySEECAiIiJyEC5KF0BERERkTQw3RERE5FAYboiIiMihMNwQERGRQ2G4ISIiIofCcENEREQOheGGiIiIHArDDRERETkUhhsiIiJyKAw3RFWUnJwMSZLMD7VajfDwcIwYMQJ///230uUBAIYPH47IyMgq7fvFF18gKSmp3NckScKsWbOqXFdFDB8+HDExMVY51s0/J0mS4OnpiaZNm2L27NkoKCiwynuU53Z9WJ6YmJgytZoev/76K2bNmgVJkiz2WbJkCZKTk61bOFENp1a6AKKabtWqVWjSpAmuX7+OXbt2ITExETt37sTRo0fh6empdHlV9sUXX+DXX39FfHx8mdf27duH8PBw+Yuqhv79++Pll18GAOTn52Pnzp2YM2cOjhw5gm+++cYm73m7PryV+vXr4/PPPy+zvUGDBhg1ahQeeeQRi+1LlixBQEAAhg8fXs1qiRwHww1RNTVr1gytW7cGAHTu3Bl6vR5z587F+vXr8fTTTytcnW08+OCDSpdQacHBwRZ1d+3aFefOncPnn3+OoqIiaLVaBav7h7u7+y37Nzw8vMaFSiIlcFqKyMpM/zCdO3cOAFBUVIRp06YhKioKbm5uqFOnDsaOHYurV69a7BcZGYnevXtj3bp1aNGiBbRaLerXr4/FixdbtDNNh509e9Zi+44dOyBJEnbs2HHb+t5//3089NBDCAoKgqenJ5o3b44FCxZAp9OZ28TExGDTpk04d+6cxdSISXnTUr/++iv69u2LWrVqQavV4t5778XHH39cbo2rV6/GjBkzEBYWBh8fH3Tt2hW//fbbbesGgKVLl6Jly5bw8vKCt7c3mjRpgunTp99xv1vx9fWFJElQqVQW23/44Qd06dIFPj4+8PDwQPv27fHjjz9atLl06RKef/55REREQKPRIDAwEO3bt8cPP/wA4M59WBX/npaKjIzEsWPHsHPnTvPxTdOQle1ra3xmAEhNTUXv3r0RFBQEjUaDsLAw9OrVC3/99Ve1PjtRZXDkhsjK/vzzTwBAYGAghBDo168ffvzxR0ybNg0dO3bEkSNHMHPmTOzbtw/79u2DRqMx75uWlob4+HjMmjULISEh+PzzzzF+/HiUlJRg4sSJVqnv1KlTGDx4sDls/fLLL3j99ddx8uRJrFy5EoBxquP555/HqVOnsG7dujse87fffkO7du0QFBSExYsXw9/fH5999hmGDx+OixcvYvLkyRbtp0+fjvbt22P58uXIzc3FlClT0KdPH5w4ccIcNP69juTLL79EXFwcxo0bh7feegsuLi74888/cfz48Qp9biEESktLAfwzLfXxxx/jqaeegqurq7ndZ599hqFDh6Jv3774+OOP4erqig8++ADdu3fH999/jy5dugAAhgwZgsOHD+P111/HXXfdhatXr+Lw4cPIycmpUh/ezFSniYuLC1xcyv5fdN26dejfvz98fX2xZMkSALD4fQIq1tfW+swFBQXo1q0boqKi8P777yM4OBiZmZnYvn078vLyKtUHRNUiiKhKVq1aJQCI/fv3C51OJ/Ly8sTGjRtFYGCg8Pb2FpmZmWLr1q0CgFiwYIHFvmvWrBEAxIcffmjeVq9ePSFJkkhLS7No261bN+Hj4yMKCgos3vfMmTMW7bZv3y4AiO3bt5u3DRs2TNSrV++Wn0Gv1wudTic++eQToVKpxOXLl82v9erV65b7AhAzZ840P3/qqaeERqMR6enpFu169OghPDw8xNWrVy1q7Nmzp0W7//73vwKA2Ldv3y1rffHFF4Wfn98tX78dAOU+evToIfLz883tCgoKRO3atUWfPn0s9tfr9aJly5bigQceMG/z8vIS8fHxt33f2/VheTp16lRunU8//bQQQoiZM2eKf/+1fc8994hOnTqVOVZF+9qan/ngwYMCgFi/fn2FPzORLXBaiqiaHnzwQbi6usLb2xu9e/dGSEgItmzZguDgYPz0008AUGax55NPPglPT88yw/733HMPWrZsabFt8ODByM3NxeHDh61Sb2pqKh599FH4+/tDpVLB1dUVQ4cOhV6vx++//16lY/7000/o0qULIiIiLLYPHz4chYWF2Ldvn8X2Rx991OJ5ixYtAPwzlVeeBx54AFevXsWgQYPw7bffIjs7u1I1DhgwAAcOHMCBAwewa9cuLF68GAcPHsQjjzyC4uJiAMDevXtx+fJlDBs2DKWlpeaHwWDAI488ggMHDpjPrnrggQeQnJyMefPmYf/+/RbTetXRoEEDc52mx9y5c6t8vDv1tTU/c8OGDVGrVi1MmTIFy5Ytq/CoGpG1cVqKqJo++eQTNG3aFGq1GsHBwQgNDTW/lpOTA7VajcDAQIt9JElCSEiIeTjfJCQkpMzxTdv+3bYq0tPT0bFjRzRu3BjvvPMOIiMjodVq8fPPP2Ps2LG4fv16lY6bk5Nj8blNwsLCzK/fzN/f3+K5aSrldu8/ZMgQlJaW4qOPPsITTzwBg8GA+++/H/PmzUO3bt3uWGNgYKB54TcAdOzYEYGBgRg0aBCSk5MxevRoXLx4EYDxzKpbuXz5Mjw9PbFmzRrMmzcPy5cvx6uvvgovLy889thjWLBgQbk/x4rSarUWdVbXnframp/Z19cXO3fuxOuvv47p06fjypUrCA0NxXPPPYdXXnnFYvqPyJYYboiqqWnTprf8x8jf3x+lpaW4dOmSRcARQiAzMxP333+/RfvMzMwyxzBtM/0jZTqrxzTaYFKRkYz169ejoKAAa9euRb169czb09LS7rjv7fj7+yMjI6PM9gsXLgAAAgICqnV8kxEjRmDEiBEoKCjArl27MHPmTPTu3Ru///67xeepKNMoxi+//GJR57vvvnvLM5aCg4PNbZOSkpCUlIT09HRs2LABU6dORVZWFrZu3VqVj6cIa3/m5s2b48svv4QQAkeOHEFycjLmzJkDd3d3TJ06VZ4PRU6P01JENmRaiPnZZ59ZbP/mm29QUFBgft3k2LFj5n9oTb744gt4e3ujVatWAGA+G+bIkSMW7TZs2HDHekxn2ty86FQIgY8++qhMW41GU+GRnC5duuCnn34yhxmTTz75BB4eHlY/ddzT0xM9evTAjBkzUFJSgmPHjlXpOKZQFxQUBABo3749/Pz8cPz4cbRu3brch5ubW5nj1K1bFy+++CK6detmMX1YmT6squq+h7U/s4kkSWjZsiUWLVoEPz8/q02rElUER26IbKhbt27o3r07pkyZgtzcXLRv3958ttR9992HIUOGWLQPCwvDo48+ilmzZiE0NBSfffYZUlJS8Oabb8LDwwMAcP/996Nx48aYOHEiSktLUatWLaxbtw579uypUD1ubm4YNGgQJk+ejKKiIixduhRXrlwp07Z58+ZYu3Ytli5diujoaLi4uNxyhGrmzJnYuHEjOnfujNdeew21a9fG559/jk2bNmHBggXw9fWtQu9Zeu655+Du7o727dsjNDQUmZmZSExMhK+vb5kRsPJcvHgR+/fvB2A8PT8tLQ3z5s2Dn58fRowYAQDw8vLCu+++i2HDhuHy5cvo378/goKCcOnSJfzyyy+4dOkSli5dimvXrqFz584YPHgwmjRpAm9vbxw4cABbt27F448/XqU+rCrTSMmaNWtQv359aLVaNG/evML7W/Mzb9y4EUuWLEG/fv1Qv359CCGwdu1aXL16tUJTh0RWo+x6ZqKay3TW0oEDB27b7vr162LKlCmiXr16wtXVVYSGhooXXnhBXLlyxaJdvXr1RK9evcTXX38t7rnnHuHm5iYiIyPFwoULyxzz999/F7GxscLHx0cEBgaKcePGiU2bNlXobKnvvvtOtGzZUmi1WlGnTh0xadIksWXLljL7Xr58WfTv31/4+fkJSZIsztLBv86WEkKIo0ePij59+ghfX1/h5uYmWrZsKVatWmXRxnQGz1dffWWx/cyZMwJAmfY3+/jjj0Xnzp1FcHCwcHNzE2FhYWLAgAHiyJEjt9zn5npvfri6uor69euLESNGiD///LNM+507d4pevXqJ2rVrC1dXV1GnTh3Rq1cvc91FRUVizJgxokWLFsLHx0e4u7uLxo0bi5kzZ5rPartTH5anU6dO4p577rnl6+WdLXX27FkRGxsrvL29BQDzz7uyfW2Nz3zy5EkxaNAg0aBBA+Hu7i58fX3FAw88IJKTk2/7uYmsTRJCCPkjFRH9W2RkJJo1a4aNGzcqXQoRUY3GNTdERETkUBhuiIiIyKFwWoqIiIgciqIjN7t27UKfPn0QFhYGSZKwfv36O+6zc+dOREdHm28quGzZMtsXSkRERDWGouGmoKAALVu2xHvvvVeh9mfOnEHPnj3RsWNHpKamYvr06XjppZfwzTff2LhSIiIiqinsZlpKkiSsW7cO/fr1u2WbKVOmYMOGDThx4oR525gxY/DLL7+UuXcNEREROacadRG/ffv2ITY21mJb9+7dsWLFCuh0unLvW1JcXGxxmXqDwYDLly/D39/ffLVWIiIism9CCOTl5SEsLAwuLrefeKpR4SYzM9N8jxOT4OBglJaWIjs7u9wb9yUmJmL27NlylUhEREQ2dP78eYSHh9+2TY0KNwDKjLaYZtVuNQozbdo0JCQkmJ9fu3YNdevWxZkzZ+Dt7W3V2nQ6HbZv347OnTvb9d1vp4y8goXfNgIA6LKygDskYHtVU/rbUbC/5cX+lhf7W15V6e+8vDxERUVV6N/uGhVuQkJCytw1OSsrC2q12nzH5H/TaDQWNwk0qV27Nnx8fKxan06ng4eHB/z9/e36D4envwbmT+7hAXh6KllOldWU/nYU7G95sb/lxf6WV1X629SuIktKatR/2du2bYuUlBSLbdu2bUPr1q35y1gJLp7u/zyx8R2LiYiI5KZouMnPz0daWhrS0tIAGE/1TktLQ3p6OgDjlNLQoUPN7ceMGYNz584hISEBJ06cwMqVK7FixQpMnDhRifJrLK2nCkW4MZpVWKhsMURERFam6LTUwYMH0blzZ/Nz09qYYcOGITk5GRkZGeagAwBRUVHYvHkzJkyYgPfffx9hYWFYvHgxnnjiCdlrr8m0WuA63KFFMcMNERE5HEXDTUxMDG53mZ3k5OQy2zp16oTDhw/bsCrH5+4OFMIDtXCV01JERFak1+uh0+mULsPu6XQ6qNVqFBUVQa/Xm7e7ubnd8TTviqhRC4rJOrRaY7gBwJEbIiIrEEIgMzMTV69eVbqUGkEIgZCQEJw/f95igbCLiwuioqLg5uZWreMz3Dghd3fjtBQAjtwQEVmBKdgEBQXBw8ODF4m9A4PBgPz8fHh5eZlHagwGAy5cuICMjAzUrVu3Wn3IcOOEOHJDRGQ9er3eHGxudVkSsmQwGFBSUgKtVmsxDRUYGIgLFy6gtLS0WmdB16hTwck6LEZuGG6IiKrFtMbGw8ND4UpqPtN01M3rcKqC4cYJWYzccFqKiMgqOBVVfdbqQ4YbJ2Q6WwoAR26IiMjhMNw4IdN1bgAw3BARkdXExMQgPj5e6TK4oNgZWYzccFqKiMjp3Gn6x3Qx3cpau3atXdwOieHGCfFsKSIi55aRkWH+fs2aNXjttdfw22+/mbe5u7tbtNfpdBUKLbVr17ZekdXAaSknxOvcEBE5t5CQEPPD19cXkiSZnxcVFcHPzw///e9/ERMTA61Wi88++ww5OTkYNGgQwsPD4eHhgebNm2P16tUWx/33tFRkZCTeeOMNPPvss/D29kbdunXx4Ycf2vzzceTGCd08LSUKCsH1/URE1iOEcoPiHh6AtU7amjJlCt5++22sWrUKGo0GRUVFiI6OxpQpU+Dj44NNmzZhyJAhqF+/Ptq0aXPL47z99tuYO3cupk+fjq+//hovvPACOnTogLCwMOsUWg6GGyd084JiQ34hVArXQ0TkSAoLAS8vZd47Px/w9LTOseLj4/H4449bbJs4caL5+3HjxmHr1q346quvbhtuevbsibi4OADGwLRo0SLs2LEDgwcPtk6h5WC4cUI3j9wYCq4z3BARURmtW7e2eK7X6zF//nysWbMGf//9N4qLi1FcXAzPO6SpFi1amL83TX9dunTJJjWbMNw4ITc34Lop3ORzQTERkTV5eBhHUJR6b2v5d2h5++23sWjRIiQlJaF58+bw9PREfHw8SkpKbnucfy9EliQJBoPBeoWWg+HGCUkSUOrmDpQAopALiomIrEmSrDc1ZE92796Nvn374plnngFgvD/UH3/8gaZNmypcWVk8W8pJlbrdWFDMU8GJiKgCGjZsiJSUFOzduxcnTpzA6NGjkZmZqXRZ5WK4cVIGDa9zQ0REFffqq6+iVatW6N69O2JiYhASEoJ+/fopXVa5OC3lpAwa49lSUhGnpYiInNnw4cMxfPhw8/PIyEgIIcq0q127NtavX3/bY+3YscPi+dmzZ8u0SUtLg8FgQG5ubhWqrRiO3Dgp4W4cuZGuc+SGiIgcC8ONkxJa48iNSxHDDRERORaGG2d143xBVQmnpYiIyLEw3DirG+HGRV8K6HQKF0NERGQ9DDdOysXzpju+8uaZRETkQBhunJTKQwOD6ZaZPB2ciIgcCMONk3L3kMz3l2K4ISIiR8Jw46Tc3f+5MzinpYiIyJEw3DgprRYcuSEiIofEcOOkLEZuGG6IiKiSYmJiEB8fr3QZ5WK4cVIWIzecliIicip9+vRB165dy31t3759kCQJhw8flrkq62G4cVLu7pyWIiJyViNHjsRPP/2Ec+fOlXlt5cqVuPfee9GqVSsFKrMOhhsnpdVyQTERkbPq3bs3goKCkJycbLG9sLAQa9asQb9+/TBo0CCEh4fDw8MDzZs3x+rVq5Uptgp4V3AnxZEbIiIbEUK5v1c9PABJumMztVqNoUOHIjk5Ga+99hqkG/t89dVXKCkpwahRo7B69WpMmTIFPj4+2LRpE4YMGYL69eujTZs2tv4U1cZw46R4thQRkY0UFgJeXsq8d34+4OlZoabPPvss/vOf/2DHjh3o3LkzAOOU1OOPP446depg4sSJ5rbjxo3D1q1b8dVXXzHckP1ydwdyOS1FROS0mjRpgnbt2mHlypXo3LkzTp06hd27d2Pbtm3Q6/WYP38+1qxZg7///hvFxcUoLi6GZwWDk9IYbpyUVgtkcuSGiMj6PDyMIyhKvXcljBw5Ei+++CLef/99rFq1CvXq1UOXLl3wn//8B4sWLUJSUhKaN28OT09PxMfHo6SkxEaFWxfDjZPiFYqJiGxEkio8NaS0AQMGYPz48fjiiy/w8ccf47nnnoMkSdi9ezf69u2LZ555BgBgMBjwxx9/oGnTpgpXXDE8W8pJcc0NERF5eXlh4MCBmD59Oi5cuIDhw4cDABo2bIiUlBTs3bsXJ06cwOjRo5GZmalssZXAcOOkeLYUEREBxqmpK1euoGvXrqhbty4A4NVXX0WrVq3QvXt3xMTEICQkBP369VO20ErgtJST4nVuiIgIANq2bQshhMW22rVrY/369bfdb8eOHbYrqpo4cuOkOHJDRESOiuHGSd285kYw3BARkQNhuHFSN58tJQo4LUVERI6D4cZJ3TwtJQo4ckNERI6D4cZJubkBRaaRm0KO3BARVde/F+VS5VmrDxlunJQkAXrNjQXF1zlyQ0RUVa6urgCMd9Sm6jFdAVmlUlXrODwV3InpNR5AMXi2FBFRNahUKvj5+SErKwsA4OHhYb7LNpXPYDCgpKQERUVFcHFxMW+7dOkSPDw8oFZXL54w3DgxoXUHcgGpiNNSRETVERISAgDmgEO3J4TA9evX4e7ubhEEXVxcULdu3WqHQ4YbJ2bQGqelVEWFgBDGuSoiIqo0SZIQGhqKoKAg6HQ6pcuxezqdDrt27cJDDz1kntYDADc3N/NITnUw3Dizm+8eW1RkPIWKiIiqTKVSVXu9iDNQqVQoLS2FVqu1CDfWwgXFTkzyuCnM8BYMRETkIBhunJibhxoluJGYuaiYiIgcBMONE+PNM4mIyBEx3Dgx3jyTiIgcEcONE7v55pkMN0RE5CgYbpzYzTfP5LQUERE5CoYbJ8aRGyIickQMN07MYs0NR26IiMhBMNw4MYtpKY7cEBGRg2C4cWKcliIiIkfEcOPEuKCYiIgcEcONE+PIDREROSKGGyfGi/gREZEjYrhxYrz9AhEROSLFw82SJUsQFRUFrVaL6Oho7N69+7btP//8c7Rs2RIeHh4IDQ3FiBEjkJOTI1O1joUjN0RE5IgUDTdr1qxBfHw8ZsyYgdTUVHTs2BE9evRAenp6ue337NmDoUOHYuTIkTh27Bi++uorHDhwAKNGjZK5csdgseaGIzdEROQgFA03CxcuxMiRIzFq1Cg0bdoUSUlJiIiIwNKlS8ttv3//fkRGRuKll15CVFQUOnTogNGjR+PgwYMyV+4YeJ0bIiJyRGql3rikpASHDh3C1KlTLbbHxsZi79695e7Trl07zJgxA5s3b0aPHj2QlZWFr7/+Gr169brl+xQXF6O4uNj8PDc3FwCg0+mg0+ms8En+YTqetY9rK2q1ZB65MRQUQF9D6japaf1d07G/5cX+lhf7W15V6e/KtFUs3GRnZ0Ov1yM4ONhie3BwMDIzM8vdp127dvj8888xcOBAFBUVobS0FI8++ijefffdW75PYmIiZs+eXWb7tm3b4OHhUb0PcQspKSk2Oa61nTrlax65yfnrL+zdvFnhiqqmpvS3o2B/y4v9LS/2t7wq09+FlZhhUCzcmEiSZPFcCFFmm8nx48fx0ksv4bXXXkP37t2RkZGBSZMmYcyYMVixYkW5+0ybNg0JCQnm57m5uYiIiEBsbCx8fHys90FgTJUpKSno1q0bXF1drXpsWzh+HNiOrQAAf3d39OzZU+GKKqem9XdNx/6WF/tbXuxveVWlv00zLxWhWLgJCAiASqUqM0qTlZVVZjTHJDExEe3bt8ekSZMAAC1atICnpyc6duyIefPmITQ0tMw+Go0GGo2mzHZXV1eb/QLb8tjW5OPzz4Jil+vX4VIDai5PTelvR8H+lhf7W17sb3lVpr8r83NRbEGxm5sboqOjywxJpaSkoF27duXuU1hYCBcXy5JVKhUA44gPVc7N17kRPFuKiIgchKJnSyUkJGD58uVYuXIlTpw4gQkTJiA9PR1jxowBYJxSGjp0qLl9nz59sHbtWixduhSnT5/G//3f/+Gll17CAw88gLCwMKU+Ro3F69wQEZEjUnTNzcCBA5GTk4M5c+YgIyMDzZo1w+bNm1GvXj0AQEZGhsU1b4YPH468vDy89957ePnll+Hn54eHH34Yb775plIfoUazDDccuSEiIseg+ILiuLg4xMXFlftacnJymW3jxo3DuHHjbFyVc3BzA4puTEtJ164CTZrcfoe77gK+/tq4IxERkZ1SPNyQciQJyNMGIrvIHwHIAX777fY7/PYbcOAA0L69PAUSERFVAcONk5PctWhadAIHPjmJyMjbNBw7Fjh6FLh4Ua7SiIiIqoThxsm5uwMXrgTiSrNARN53m4b16xvDTVaWbLURERFVheJ3BSdlabXGr3c8E9x07SGGGyIisnMMN07O/cZ9M4uK7tAwKMj4leGGiIjsHMONk6vwyA3DDRER1RAMN06u0iM3XFBMRER2juHGyXHkhoiIHA3DjZMzjdxwQTERETkKhhsnZxq5qfC01OXLgE5n05qIiIiqg+HGyVV45KZ2bcB0R/bsbJvWREREVB0MN06uwiM3Li5AYKDxey4qJiIiO8Zw4+QqPHIDcFExERHVCAw3Tq7CIzcAFxUTEVGNwHDj5DhyQ0REjobhxslV+CJ+AMMNERHVCAw3Tq7CF/EDeJViIiKqERhunBxHboiIyNEw3Di5So3ccEExERHVAAw3To4jN0RE5GgYbpxcldbcZGUBQtisJiIioupguHFylRq5MV2huKgIyMuzWU1ERETVwXDj5Co1cuPpaXwAnJoiIiK7xXDj5Cp1ET+Ai4qJiMjuMdw4uUrdfgHgomIiIrJ7aqULIGWZRm4KCoCdO2/fNjISqMcL+RERkZ1juHFyHh7Gr8XFQEzM7du6ugJX+gfBE+DIDRER2S2GGycXFASMGgXs2XP7dmfPGqeuMgxBaAgw3BARkd1iuHFykgR89NGd2/XuDWzaBGQhmOGGiIjsGhcUU4WEhxu//q3jgmIiIrJvDDdUIaZwc7aQC4qJiMi+MdxQhZjCzR/XOHJDRET2jeGGKiQiwvj1RM6NcJOTA5SWKlcQERHRLTDcUIWYRm5+zfAHXG782mRnK1cQERHRLTDcUIXUqWP8ejVPBYN/gPEJp6aIiMgOMdxQhXh5AX5+xu9L/LiomIiI7BfDDVWYaWqqwJOLiomIyH4x3FCFmcLNNTeGGyIisl8MN1RhpnCT5RJ84xuGGyIisj8MN1RhpnCTUcqRGyIisl8MN1RhpmvdpBdxQTEREdkvhhuqMNPIzak8jtwQEZH9YrihCjOFm5OXGW6IiMh+MdxQhZnvL5V304JiIZQriIiIqBwMN1RhPj6AtzeQhRsjN9evAwUFyhZFRET0Lww3VCnh4UAhPKHXehg3cFExERHZGYYbqhTT1FShF9fdEBGRfWK4oUoxnQ6eq2W4ISIi+8RwQ5ViGrnJUfEqxUREZJ8YbqhSTOEm08CRGyIisk8MN1QppnBzvphXKSYiIvvEcEOVYgo3p/M5ckNERPaJ4YYqxRRuzhQy3BARkX1iuKFK8fMDPDyAi+CCYiIisk8MN1QpkmQ8Hdx8lWKGGyIisjMMN1Rp4eE3hZvsbKC0VNmCiIiIbsJwQ5UWHg7kwB9Ckow3zszJUbokIiIiM4YbqrTwcEAPNfI1/sYNnJoiIiI7wnBDlWY6Y+qKK9fdEBGR/WG4oUozhZuL4sYZU7yQHxER2RGGG6o0U7j5W8eRGyIisj8MN1RppnCTXsxwQ0RE9ofhhirN3x/QanmtGyIisk+Kh5slS5YgKioKWq0W0dHR2L17923bFxcXY8aMGahXrx40Gg0aNGiAlStXylQtAcYL+Vlc64bhhoiI7IhayTdfs2YN4uPjsWTJErRv3x4ffPABevTogePHj6Nu3brl7jNgwABcvHgRK1asQMOGDZGVlYVSXkROduHhwMU/uaCYiIjsj6LhZuHChRg5ciRGjRoFAEhKSsL333+PpUuXIjExsUz7rVu3YufOnTh9+jRq164NAIiMjJSzZLohPBz4kyM3RERkhxQLNyUlJTh06BCmTp1qsT02NhZ79+4td58NGzagdevWWLBgAT799FN4enri0Ucfxdy5c+Hu7l7uPsXFxSguLjY/z83NBQDodDrodDorfRqYj3nzV0cWGuqCvTfCjcjKQqkCn9mZ+tsesL/lxf6WF/tbXlXp78q0VSzcZGdnQ6/XIzg42GJ7cHAwMjMzy93n9OnT2LNnD7RaLdatW4fs7GzExcXh8uXLt1x3k5iYiNmzZ5fZvm3bNnh4eFT/g5QjJSXFJse1J9euRSELUQAAqbAQ369dC71Wq0gtztDf9oT9LS/2t7zY3/KqTH8XFhZWuK2i01IAIEmSxXMhRJltJgaDAZIk4fPPP4evry8A49RW//798f7775c7ejNt2jQkJCSYn+fm5iIiIgKxsbHw8fGx4icxpsqUlBR069YNrq6uVj22vSktlfDhhyoUSVpoRRG633cfEBUlaw3O1N/2gP0tL/a3vNjf8qpKf5tmXipCsXATEBAAlUpVZpQmKyurzGiOSWhoKOrUqWMONgDQtGlTCCHw119/oVGjRmX20Wg00Gg0Zba7urra7BfYlse2F6Ycc8klGBH6c3C9fBm46y5FanGG/rYn7G95sb/lxf6WV2X6uzI/F8VOBXdzc0N0dHSZIamUlBS0a9eu3H3at2+PCxcuID8/37zt999/h4uLC8JNV5YjWZi6O0PPRcVERGRfFL3OTUJCApYvX46VK1fixIkTmDBhAtLT0zFmzBgAximloUOHmtsPHjwY/v7+GDFiBI4fP45du3Zh0qRJePbZZ2+5oJhsIyAAcHPjtW6IiMj+KLrmZuDAgcjJycGcOXOQkZGBZs2aYfPmzahXrx4AICMjA+np6eb2Xl5eSElJwbhx49C6dWv4+/tjwIABmDdvnlIfwWm5uAB16gBZZxhuiIjIvii+oDguLg5xcXHlvpacnFxmW5MmTbia3U6EhzPcEBGR/VH89gtUc4WHAxfBqxQTEZF9YbihKuP9pYiIyB4x3FCVMdwQEZE9YrihKouIYLghIiL7w3BDVWYxcpOdDej1yhZEREQEhhuqhvBw4BICjU8MBiAnR9mCiIiIwHBD1RAUBEhqNbLhb9zAqSkiIrIDDDdUZSoVEBbGdTdERGRfGG6oWnjGFBER2RurhBu9Xo+0tDRcuXLFGoejGoThhoiI7E2Vwk18fDxWrFgBwBhsOnXqhFatWiEiIgI7duywZn1k5yIieJViIiKyL1UKN19//TVatmwJAPjuu+9w5swZnDx5EvHx8ZgxY4ZVCyT7xpEbIiKyN1UKN9nZ2QgJCQEAbN68GU8++STuuusujBw5EkePHrVqgWTfGG6IiMjeVCncBAcH4/jx49Dr9di6dSu6du0KACgsLIRKpbJqgWTfGG6IiMjeqKuy04gRIzBgwACEhoZCkiR069YNAPC///0PTZo0sWqBZN9uDjciKwuSwvUQERFVKdzMmjULzZo1w/nz5/Hkk09Co9EAAFQqFaZOnWrVAsm+hYQAl6RgQAAi8yLDDRERKa5K4QYA+vfvb/H86tWrGDZsWLULoppFrQZcQoKADMClsAAoKAA8PZUui4iInFiV1ty8+eabWLNmjfn5gAED4O/vj/DwcBw5csRqxVHN4BfhjSIYR+9w6ZKyxRARkdOrUrj54IMPEBERAQBISUlBSkoKtmzZgkceeQQTJ060aoFk/yLqSlxUTEREdqNK01IZGRnmcLNx40YMGDAAsbGxiIyMRJs2baxaINk/06LiujjPC/kREZHiqjRyU6tWLZw/fx4ALE4FF0JAr9dbrzqqEcLDb7pKMUduiIhIYVUauXn88ccxePBgNGrUCDk5OejRowcAIC0tDQ0bNrRqgWT/eK0bIiKyJ1UKN4sWLUJkZCTOnz+PBQsWwMvLC4BxuiouLs6qBZL9Cw8H9jDcEBGRnahSuHF1dS134XB8fHx166EayOJCfhd5IT8iIlJWldbcAMCnn36KDh06ICwsDOfOnQMAJCUl4dtvv7VacVQzhIUBl26EG91fXFBMRETKqlK4Wbp0KRISEtCjRw9cvXrVvIjYz88PSUlJ1qyPagBXV0BX27igWJ/BaSkiIlJWlcLNu+++i48++ggzZsywuFFm69ateVdwJ+USahy5kbIZboiISFlVCjdnzpzBfffdV2a7RqNBQUFBtYuimkcbYQw3btcuAQaDwtUQEZEzq1K4iYqKQlpaWpntW7Zswd13313dmqgG8qkfAABwEQbg8mWFqyEiImdWpbOlJk2ahLFjx6KoqAhCCPz8889YvXo1EhMTsXz5cmvXSDVAaD03XEYt1MYV41WKAwKULomIiJxUlcLNiBEjUFpaismTJ6OwsBCDBw9GnTp18M477+Cpp56ydo1UA5iuUlwbV4zXurnnHqVLIiIiJ1WlcAMAzz33HJ577jlkZ2fDYDAgKCjImnVRDWO61k1TnOSF/IiISFFVDjcmAZx+IBjDzSHThfwyL/JCfkREpJgqLSi+ePEihgwZgrCwMKjVaqhUKosHOZ86df65SvH1dI7cEBGRcqo0cjN8+HCkp6fj1VdfRWhoKCSJ/093dhoNkO8ZDBQA189chIfSBRERkdOqUrjZs2cPdu/ejXvvvdfK5VBNVlrbGG50FzhyQ0REyqnStFRERASEENauhWo4l5AbVynO4v2liIhIOVUKN0lJSZg6dSrOnj1r5XKoJlOFGe8v5XaF4YaIiJRTpWmpgQMHorCwEA0aNICHhwdcXV0tXr/MK9Q6JW09Y7jxyOe0FBERKadK4WbRokVcRExleDcwTktpSguB/HzAy0vhioiIyBlV+Wwpon8LiPRCIdzhgevGWzAw3BARkQKqtOZGpVIhq5yr0Obk5PA6N04sJFTCRRinpniVYiIiUkqVws2tzpQqLi6Gm5tbtQqimis09J8L+RkyuKiYiIiUUalpqcWLFwMAJEnC8uXL4XXTtINer8euXbvQpEkT61ZINUZwMHD4xshNwemL8Fa4HiIick6VCjeLFi0CYBy5WbZsmcUUlJubGyIjI7Fs2TLrVkg1hqsrcE0bDBQBBWeyGG6IiEgRlQo3Z86cAQB07twZa9euRa1atWxSFNVcRT5BQBFQcp7TUkREpIwqnS21fft2a9dBDqK0djCQxTU3RESknAqHm4SEBMydOxeenp5ISEi4bduFCxdWuzCqoYKCgJOASzbPliIiImVUONykpqbi5MmTuO+++5CamnrLdry4n3NT1zEuKNZc5cgNEREpo8LhZvv27VCpVMjIyDBPSw0cOBCLFy9GcHCwzQqkmsU90vi74FnAkRsiIlJGpa5z8+/r22zZsgUFBQVWLYhqNtMtGLx0V4CSEoWrISIiZ1Sli/iZ3OpifuS8/BvVRiluXCKAVykmIiIFVCrcSJJUZk0N19jQzULruJivUiwuMtwQEZH8KnUquBACw4cPh0ajAQAUFRVhzJgx8PT0tGi3du1a61VINUpoKPA7ghCGDFw/exEe0UpXREREzqZS4WbYsGEWz5955hmrFkM1n4cHkKMKBvTAtd8vwkPpgoiIyOlUKtysWrXKVnWQA8nzCAbygMKznJYiIiL5VWtBMVF5in2Na250f/FaN0REJD+GG7K6Un/jtW5EJsMNERHJj+GGrE4KMYYbVQ6npYiISH4MN2R1ruHGaSntNY7cEBGR/BhuyOo8btyCwauQ4YaIiOTHcENW59vIOHLjW3IJMBgUroaIiJwNww1ZXa3GxnCjggHIyVG4GiIicjaKh5slS5YgKioKWq0W0dHR2L17d4X2+7//+z+o1Wrce++9ti2QKi20rityUBsAUHyei4qJiEheioabNWvWID4+HjNmzEBqaio6duyIHj16ID09/bb7Xbt2DUOHDkWXLl1kqpQqo1Yt4JJkHL25cpLrboiISF6KhpuFCxdi5MiRGDVqFJo2bYqkpCRERERg6dKlt91v9OjRGDx4MNq2bStTpVQZkgRcdTMuKs79g+GGiIjkVanbL1hTSUkJDh06hKlTp1psj42Nxd69e2+536pVq3Dq1Cl89tlnmDdv3h3fp7i4GMXFxebnubm5AACdTgedTlfF6stnOp61j1sT5XsGAcVAwdkMm/UH+1te7G95sb/lxf6WV1X6uzJtFQs32dnZ0Ov1CA4OttgeHByMzMzMcvf5448/MHXqVOzevRtqdcVKT0xMxOzZs8ts37ZtGzw8bHNbx5SUFJsctyYpcfUCAGSmHUP65s02fS/2t7zY3/Jif8uL/S2vyvR3YWFhhdsqFm5MJEmyeC6EKLMNAPR6PQYPHozZs2fjrrvuqvDxp02bhoSEBPPz3NxcREREIDY2Fj4+PlUvvBw6nQ4pKSno1q0bXF1drXrsmmZL3V+Ai0CgQaBlz542eQ/2t7zY3/Jif8uL/S2vqvS3aealIhQLNwEBAVCpVGVGabKyssqM5gBAXl4eDh48iNTUVLz44osAAIPBACEE1Go1tm3bhocffrjMfhqNBhqNpsx2V1dXm/0C2/LYNYVLaCgAwPXKJZv3BftbXuxvebG/5cX+lldl+rsyPxfFFhS7ubkhOjq6zJBUSkoK2rVrV6a9j48Pjh49irS0NPNjzJgxaNy4MdLS0tCmTRu5SqcKcIu4cQuGXC4oJiIieSk6LZWQkIAhQ4agdevWaNu2LT788EOkp6djzJgxAIxTSn///Tc++eQTuLi4oFmzZhb7BwUFQavVltlOyvOMMo6++fAWDEREJDNFw83AgQORk5ODOXPmICMjA82aNcPmzZtRr149AEBGRsYdr3lD9sn3LmO48dNlAUIYzw8nIiKSgeILiuPi4hAXF1fua8nJybfdd9asWZg1a5b1i6JqC7jbOC3ljiLor+ZBVcu6i7eJiIhuRfHbL5BjCoz0RD48AQCXT/IWDEREJB+GG7IJlQrIVhmnpngLBiIikhPDDdlMrsY4NZX3J8MNERHJh+GGbKbAyzhyU5TOaSkiIpIPww3ZTImfceSm9G+O3BARkXwYbshmDIHGkRspi+GGiIjkw3BDNqMKM4Yb1yucliIiIvkw3JDNaG7cgsE9jyM3REQkH4YbshmvBsaRG9/rDDdERCQfhhuyGb/GxnBTqzQLQihcDBEROQ2GG7IZ/6bGaSk/XMPVzCKFqyEiImfBcEM2ow2thRK4AgAuHeOiYiIikgfDDdmOJOGy2jh6c/V3hhsiIpIHww3ZVJ7WGG4KTnNRMRERyYPhhmyq0Nu4qLj4PEduiIhIHgw3ZFO62sZwo7/AkRsiIpIHww3ZlAg0TktJlxhuiIhIHgw3ZFOmWzC48RYMREQkE4YbsiltXePIjUc+R26IiEgeDDdkU14Nb9yCoYjhhoiI5MFwQzZV68YtGAIMWSgsVLgYIiJyCgw3ZFNe9Y3TUgHIRubfeoWrISIiZ8BwQzYlBQXCAAkqGJB9MlvpcoiIyAkw3JBtqdW4pvYHwFswEBGRPBhuyOby3Y1TU4VnuKiYiIhsj+GGbO66z41bMKQz3BARke0x3JDN6fyN4caQyWkpIiKyPYYbsjkpyDgt5ZLNkRsiIrI9hhuyOXUd48iN5ipHboiIyPYYbsjm3CON4cazgCM3RERkeww3ZHPeDYzTUrVKLqK0VOFiiIjI4THckM35NDKO3AQhCxc5eENERDbGcEM25xJiHLkJxkVkXBAKV0NERI6O4YZsL/jGgmKU4NKf1xQuhoiIHB3DDdmeuzsK1d4AgNw/ecYUERHZFsMNySLfg7dgICIieTDckCyKbtyCQfcXww0REdkWww3JojTAGG7ERU5LERGRbTHckCxMt2BQ8RYMRERkYww3JAvXCOPIjfYaww0REdkWww3JwuPGLRi8CrMgeKkbIiKyIYYbkoV3Q+O0VKC4iJwchYshIiKHxnBDsnCt888tGDIyFC6GiIgcGsMNyePGVYqDcZHhhoiIbIrhhuRx42wpH+ThUvp1hYshIiJHxnBD8vD1hc7FDQBvwUBERLbFcEPykCQUeBpHb66f5engRERkOww3JJti3xu3YPib4YaIiGyH4YZkow80hhvwFgxERGRDDDckG5dg47SUOocjN0REZDsMNyQbtxu3YHDPZbghIiLbYbgh2XjWN4abWqVZyM9XuBgiInJYDDckG02EcVqKF/IjIiJbYrgh+QTzFgxERGR7DDckH96CgYiIZMBwQ/K5cQsGf+Qg869ShYshIiJHxXBD8gkIgEFygQsEck9nK10NERE5KIYbko9KhSKvAADAwU0XUcrBGyIisgGGG5KV6Yyp6+cuIjlZ2VqIiMgxMdyQrFSh/5wxNXMmUFiocEFERORwGG5IXjcWFTepdREXLgDvvqtwPURE5HAYbkheN04Hf6yt8RYM8+cDly8rWRARETkahhuSV1QUAODu81vRopkBV68aAw4REZG1KB5ulixZgqioKGi1WkRHR2P37t23bLt27Vp069YNgYGB8PHxQdu2bfH999/LWC1V25AhgK8vpKNHkdxjDQDj1NRffylcFxEROQxFw82aNWsQHx+PGTNmIDU1FR07dkSPHj2Qnp5ebvtdu3ahW7du2Lx5Mw4dOoTOnTujT58+SE1NlblyqrJatYBJkwAA9657DZ076FBUBMyapWxZRETkOBQNNwsXLsTIkSMxatQoNG3aFElJSYiIiMDSpUvLbZ+UlITJkyfj/vvvR6NGjfDGG2+gUaNG+O6772SunKpl/HggMBDSn39ieYdkAMCqVcCJE8qWRUREjkGt1BuXlJTg0KFDmDp1qsX22NhY7N27t0LHMBgMyMvLQ+3atW/Zpri4GMXFxebnubm5AACdTgedTleFym/NdDxrH9fhaDRwmToVqpdfRtRnc/BEr6fxzSYPTJ1qwNdf6yt8GPa3vNjf8mJ/y4v9La+q9Hdl2ioWbrKzs6HX6xF84+wZk+DgYGRmZlboGG+//TYKCgowYMCAW7ZJTEzE7Nmzy2zftm0bPDw8Kld0BaWkpNjkuI7EJSICXf394f7XXxj/wESsc3kfGza4YOHCPWjS5EqljsX+lhf7W17sb3mxv+VVmf4urMSF0RQLNyaSJFk8F0KU2Vae1atXY9asWfj2228RdOPaKeWZNm0aEhISzM9zc3MRERGB2NhY+Pj4VL3wcuh0OqSkpKBbt25wdXW16rEdkXT5MjBmDDrs/hqjBydi6We+2LixAyZM0KMCvwLsb5mxv+XF/pYX+1teVelv08xLRSgWbgICAqBSqcqM0mRlZZUZzfm3NWvWYOTIkfjqq6/QtWvX27bVaDTQaDRltru6utrsF9iWx3Yozz4LvP02pD/+QGLIu1ilfQV79rggJcUFvXpV/DDsb3mxv+XF/pYX+1telenvyvxcFFtQ7Obmhujo6DJDUikpKWjXrt0t91u9ejWGDx+OL774Ar0q8y8g2R9XV2DOHACA74f/weRRxqv5TZsG6Cu+9IaIiMiComdLJSQkYPny5Vi5ciVOnDiBCRMmID09HWPGjAFgnFIaOnSouf3q1asxdOhQvP3223jwwQeRmZmJzMxMXLt2TamPQNU1YADQogWQm4upLgvg5wccPQp88YXShRERUU2laLgZOHAgkpKSMGfOHNx7773YtWsXNm/ejHr16gEAMjIyLK5588EHH6C0tBRjx45FaGio+TF+/HilPgJVl4sLMG8eAMD9o8WYO9Y4Tfnqq8BNJ7kRERFVmOILiuPi4hAXF1fua8nJyRbPd+zYYfuCSH69ewMPPgjs34/ROa8jMexdnDsHPPEEULfurXczGFzg7R2Gnj3lK5WIiOyf4uGGCJIEvPEG8PDDcF3xAd6e8zIGTYvEpk132lEFF5doTJigR1iYHIUSEVFNwHBD9qFzZ6BLF+DHHzHw5GzkfrAKGRm33+XjjwXOnHHB998bMGKEPGUSEZH9Y7gh+/H668CPP0L69BM8P2Uy0LTpbZsXFRkwf74KW7a4MNwQEZGZ4ncFJzJr0wbo2xcwGICZM+/YvGdPAQDYtk0Cr5hOREQmHLkh+zJ3LrBhA/DVV8D27UDjxrdsen8dHWp5B+DKNU/s3Qt06iRjnUREZLcYbsi+NG8ODB4MfP458PDDt22qBXDQszEa4Rg2blQx3BAREQBOS5E9mjsXiIgAVKrbPwDUL/gND2FXBc6sIiIiZ8FwQ/YnKgpITwdKS2/7MAwfDgAYKK3BiRPA6dPKlk1ERPaB4YZqLEP//gCAAep1UKGUozdERASA4YZqMNG5M0q8vVFbl4WHsAsbNypdERER2QOGG6q5XF2R0aYNAGAA/osdO4D8fGVLIiIi5THcUI32d/v2AIAnXdZCX1KKH35QuCAiIlIcww3VaNnNm0P4+8PfcAmdsJPrboiIiOGGajahVkP07QsAeBJfYdMmQAiFiyIiIkUx3FCNZzprqj++QVZGKVJTFS6IiIgUxXBDNZ6IiQH8/RGAbMRgB6emiIicHMMN1XxqNfD44wCMU1M8JZyIyLkx3JBjGDAAAPAEvsHhn0tx8aLC9RARkWIYbsgxxMQAAQEIQA5isB1btihdEBERKYXhhhzDv6amuO6GiMh5MdyQ47gxNfU41uLHrTqUlChcDxERKYLhhhxHp04QgYEIQA5a52/Hnj1KF0REREpguCHHoVZD4llTREROj+GGHMtNU1Pfb9QpXAwRESmB4YYcy0MPwRAYBH9cRvgfP+GPP5QuiIiI5MZwQ45FrYbLEzxriojImTHckOO5MTX1GNZhywZOTRERORuGG3I8Dz2EUn/j1JR610/IzVW6ICIikhPDDTkelQrqAU8AAB7X/xcpKQrXQ0REsmK4Icd009TU1u84NUVE5EzUShdAZBMdO6K4VjBqX7mIvPU/YuPGR+DjA/j4AL6+MH/v6qp0oUREZG0cuSHHdNPUVOy1/6JPH6BTJ+C++4D69YGAAMDNDXB3B4KDgTZtgNRUhWsmIiKr4MgNOSzVoAHAB0swQL0WIrAOLun8cLGkFjKKaiGzpBauoBauFBkfP2d5oVcvCf/7HxARoXTlRERUHQw35Lg6dADCwuB14QJGZsy7bdM8Fx/0yNiE3r07YM8ewNtbphqJiMjqGG7IcalUwHffAevWAVeu3Pqh08HbkIsP1GPR4shhPPWUCt9+C6j5p4OIqEbiX9/k2Fq1Mj5uRQggMxO4+27cc/UIRqtXYunm5/Dyy8A778hXJhERWQ8XFJNzkyQgNBSYORMAsNBjBnxwDYsXA++9p3BtRERUJQw3RAAQFwfcdRe0uZfw/UNvAADGjwc2b1a4LiIiqjSGGyLAeF74228DANrsT8Lk/qdhMAADBwJHjihcGxERVQrDDZFJr15A166QSkqQqJ+Mzp2B/Hygd28gI0Pp4oiIqKIYbohMJAlYuBBwcYHLum/w7cu70LgxcP480KcPUFCgdIFERFQRPFuK6GbNmwPPPw8sWwbvV+OxacMBtGmnwqFDwJAhwIoVxgzkjHQ6ID9fjatXedsKObC/rUuSABcX4xUibv7q4uK8f6YdGcMN0b/NmQN88QWQmooG//cJ1q8fgS5djJfLWbdO6eKU5Aqgl9JFOBH2t1wkCVCp1AD6wMWFSccaQkOBs2eVe3+GG6J/CwwEXn0VmDQJmD4dHf54Ep984oXRo4Fr15QujoisTQigtFQCwGBjLSUlyr4/ww1RecaNA5YtA06dAubPx8B589C/P6DXK12YcnQ6HbZs2YIePXrAlfMkNsf+ti4hjH9+DQbjw/S96WtxsQ4//PATHn74Yfa3FbgovKKX4YaoPBoN8NZbwGOPGb8+9xxU9epBpVK6MOVIEuDqKuDmxjUgcmB/y0unAwICihAezv52BDxbiuhW+vYFYmKA4mJg6lSlqyEiogpiuCG6FUkCFi0yfv3yS2DvXqUrIiKiCmC4Ibqde+8FRo40fh8fb5ycJyIiu8Y1N0R3MneuceTmwAHgnnuM63GckFoIxOTmQv3aa/JcGESIim1zUGohEJOXB/Urr/BCLNUlhPE/JqWlxodeX+Z7dWkpepWWQuXMC+usKSQE+PNPxd6e4YboTkJCgNdeAyZPBk6eVLoaxUgAfJUuwomwv+Ulgf8gWlVhoaJvz58lUUVMnAi0b+/U92AoLS3Fzz//jAceeABqtUx/dZQ3YuEkoxilpaX43//+hzZt2sjX345MpTI+1Op/Hjc91xkM2LFrF2JiYngquDUoPALGPzFEFSFJQLt2SlehKKHT4VJJCUTXrjxXVgZCp0N2URHEww+zv+Wg06Hwt9+AqCj2twPggmIiIiJyKAw3RERE5FAYboiIiMihMNwQERGRQ2G4ISIiIofCcENEREQOheGGiIiIHArDDRERETkUhhsiIiJyKIqHmyVLliAqKgparRbR0dHYvXv3bdvv3LkT0dHR0Gq1qF+/PpYtWyZTpURERFQTKBpu1qxZg/j4eMyYMQOpqano2LEjevTogfT09HLbnzlzBj179kTHjh2RmpqK6dOn46WXXsI333wjc+VERERkrxQNNwsXLsTIkSMxatQoNG3aFElJSYiIiMDSpUvLbb9s2TLUrVsXSUlJaNq0KUaNGoVnn30Wb731lsyVExERkb1SLNyUlJTg0KFDiI2NtdgeGxuLvXv3lrvPvn37yrTv3r07Dh48CJ1OZ7NaiYiIqOZQ7K7g2dnZ0Ov1CA4OttgeHByMzMzMcvfJzMwst31paSmys7MRGhpaZp/i4mIUFxebn1+7dg0AcPnyZasHIp1Oh8LCQuTk5MCVd5W1Ofa3vNjf8mJ/y4v9La+q9HdeXh4AQAhxx7aKhRsTSZIsngshymy7U/vytpskJiZi9uzZZbZHRUVVtlQiIiJSWF5eHnx9fW/bRrFwExAQAJVKVWaUJisrq8zojElISEi57dVqNfz9/cvdZ9q0aUhISDA/NxgMuHz5Mvz9/W8boqoiNzcXEREROH/+PHx8fKx6bCqL/S0v9re82N/yYn/Lqyr9LYRAXl4ewsLC7thWsXDj5uaG6OhopKSk4LHHHjNvT0lJQd++fcvdp23btvjuu+8stm3btg2tW7e+5bCWRqOBRqOx2Obn51e94u/Ax8eHfzhkxP6WF/tbXuxvebG/5VXZ/r7TiI2JomdLJSQkYPny5Vi5ciVOnDiBCRMmID09HWPGjAFgHHUZOnSouf2YMWNw7tw5JCQk4MSJE1i5ciVWrFiBiRMnKvURiIiIyM4ouuZm4MCByMnJwZw5c5CRkYFmzZph8+bNqFevHgAgIyPD4po3UVFR2Lx5MyZMmID3338fYWFhWLx4MZ544gmlPgIRERHZGcUXFMfFxSEuLq7c15KTk8ts69SpEw4fPmzjqqpGo9Fg5syZZabByDbY3/Jif8uL/S0v9re8bN3fkqjIOVVERERENYTi95YiIiIisiaGGyIiInIoDDdERETkUBhuiIiIyKEw3FjJkiVLEBUVBa1Wi+joaOzevVvpkhzGrl270KdPH4SFhUGSJKxfv97idSEEZs2ahbCwMLi7uyMmJgbHjh1TptgaLjExEffffz+8vb0RFBSEfv364bfffrNow/62nqVLl6JFixbmC5m1bdsWW7ZsMb/OvratxMRESJKE+Ph48zb2ufXMmjULkiRZPEJCQsyv27KvGW6sYM2aNYiPj8eMGTOQmpqKjh07okePHhbX6KGqKygoQMuWLfHee++V+/qCBQuwcOFCvPfeezhw4ABCQkLQrVs3803WqOJ27tyJsWPHYv/+/UhJSUFpaSliY2NRUFBgbsP+tp7w8HDMnz8fBw8exMGDB/Hwww+jb9++5r/g2de2c+DAAXz44Ydo0aKFxXb2uXXdc889yMjIMD+OHj1qfs2mfS2o2h544AExZswYi21NmjQRU6dOVagixwVArFu3zvzcYDCIkJAQMX/+fPO2oqIi4evrK5YtW6ZAhY4lKytLABA7d+4UQrC/5VCrVi2xfPly9rUN5eXliUaNGomUlBTRqVMnMX78eCEEf7+tbebMmaJly5blvmbrvubITTWVlJTg0KFDiI2NtdgeGxuLvXv3KlSV8zhz5gwyMzMt+l+j0aBTp07sfyu4du0aAKB27doA2N+2pNfr8eWXX6KgoABt27ZlX9vQ2LFj0atXL3Tt2tViO/vc+v744w+EhYUhKioKTz31FE6fPg3A9n2t+BWKa7rs7Gzo9foydzIPDg4ucwdzsj5TH5fX/+fOnVOiJIchhEBCQgI6dOiAZs2aAWB/28LRo0fRtm1bFBUVwcvLC+vWrcPdd99t/guefW1dX375JQ4fPowDBw6UeY2/39bVpk0bfPLJJ7jrrrtw8eJFzJs3D+3atcOxY8ds3tcMN1YiSZLFcyFEmW1kO+x/63vxxRdx5MgR7Nmzp8xr7G/rady4MdLS0nD16lV88803GDZsGHbu3Gl+nX1tPefPn8f48eOxbds2aLXaW7Zjn1tHjx49zN83b94cbdu2RYMGDfDxxx/jwQcfBGC7vua0VDUFBARApVKVGaXJysoqk0jJ+kwr79n/1jVu3Dhs2LAB27dvR3h4uHk7+9v63Nzc0LBhQ7Ru3RqJiYlo2bIl3nnnHfa1DRw6dAhZWVmIjo6GWq2GWq3Gzp07sXjxYqjVanO/ss9tw9PTE82bN8cff/xh899vhptqcnNzQ3R0NFJSUiy2p6SkoF27dgpV5TyioqIQEhJi0f8lJSXYuXMn+78KhBB48cUXsXbtWvz000+IioqyeJ39bXtCCBQXF7OvbaBLly44evQo0tLSzI/WrVvj6aefRlpaGurXr88+t6Hi4mKcOHECoaGhtv/9rvaSZBJffvmlcHV1FStWrBDHjx8X8fHxwtPTU5w9e1bp0hxCXl6eSE1NFampqQKAWLhwoUhNTRXnzp0TQggxf/584evrK9auXSuOHj0qBg0aJEJDQ0Vubq7Cldc8L7zwgvD19RU7duwQGRkZ5kdhYaG5DfvbeqZNmyZ27dolzpw5I44cOSKmT58uXFxcxLZt24QQ7Gs53Hy2lBDsc2t6+eWXxY4dO8Tp06fF/v37Re/evYW3t7f530Zb9jXDjZW8//77ol69esLNzU20atXKfOosVd/27dsFgDKPYcOGCSGMpxTOnDlThISECI1GIx566CFx9OhRZYuuocrrZwBi1apV5jbsb+t59tlnzX9vBAYGii5dupiDjRDsazn8O9ywz61n4MCBIjQ0VLi6uoqwsDDx+OOPi2PHjplft2VfS0IIUf3xHyIiIiL7wDU3RERE5FAYboiIiMihMNwQERGRQ2G4ISIiIofCcENEREQOheGGiIiIHArDDRERETkUhhsiohuSk5Ph5+endBlEVE0MN0RUaZmZmRg/fjwaNmwIrVaL4OBgdOjQAcuWLUNhYaHS5VVIZGQkkpKSLLYNHDgQv//+uzIFEZHVqJUugIhqltOnT6N9+/bw8/PDG2+8gebNm6O0tBS///47Vq5cibCwMDz66KOK1CaEgF6vh1pdtb/a3N3d4e7ubuWqiEhuHLkhokqJi4uDWq3GwYMHMWDAADRt2hTNmzfHE088gU2bNqFPnz4AgGvXruH5559HUFAQfHx88PDDD+OXX34xH2fWrFm499578emnnyIyMhK+vr546qmnkJeXZ24jhMCCBQtQv359uLu7o2XLlvj666/Nr+/YsQOSJOH7779H69atodFosHv3bpw6dQp9+/ZFcHAwvLy8cP/99+OHH34w7xcTE4Nz585hwoQJkCQJkiQBKH9aaunSpWjQoAHc3NzQuHFjfPrppxavS5KE5cuX47HHHoOHhwcaNWqEDRs2mF+/cuUKnn76aQQGBsLd3R2NGjXCqlWrqv+DIKJbYrghogrLycnBtm3bMHbsWHh6epbbRpIkCCHQq1cvZGZmYvPmzTh06BBatWqFLl264PLly+a2p06dwvr167Fx40Zs3LgRO3fuxPz5882vv/LKK1i1ahWWLl2KY8eOYcKECXjmmWewc+dOi/ecPHkyEhMTceLECbRo0QL5+fno2bMnfvjhB6SmpqJ79+7o06cP0tPTAQBr165FeHg45syZg4yMDGRkZJT7WdatW4fx48fj5Zdfxq+//orRo0djxIgR2L59u0W72bNnY8CAAThy5Ah69uyJp59+2vw5X331VRw/fhxbtmzBiRMnsHTpUgQEBFS+84mo4qxy+00icgr79+8XAMTatWsttvv7+wtPT0/h6ekpJk+eLH788Ufh4+MjioqKLNo1aNBAfPDBB0IIIWbOnCk8PDxEbm6u+fVJkyaJNm3aCCGEyM/PF1qtVuzdu9fiGCNHjhSDBg0SQvxzx/j169ffsfa7775bvPvuu+bn9erVE4sWLbJos2rVKuHr62t+3q5dO/Hcc89ZtHnyySdFz549zc8BiFdeecX8PD8/X0iSJLZs2SKEEKJPnz5ixIgRd6yPiKyHa26IqNJM0zgmP//8MwwGA55++mkUFxfj0KFDyM/Ph7+/v0W769ev49SpU+bnkZGR8Pb2Nj8PDQ1FVlYWAOD48eMoKipCt27dLI5RUlKC++67z2Jb69atLZ4XFBRg9uzZ2LhxIy5cuIDS0lJcv37dPHJTUSdOnMDzzz9vsa19+/Z45513LLa1aNHC/L2npye8vb3Nn+OFF17AE088gcOHDyM2Nhb9+vVDu3btKlUHEVUOww0RVVjDhg0hSRJOnjxpsb1+/foAYF6MazAYEBoaih07dpQ5xs1rWlxdXS1ekyQJBoPBfAwA2LRpE+rUqWPRTqPRWDz/9xTZpEmT8P333+Ott95Cw4YN4e7ujv79+6OkpKSCn9SyppsJIcpsu93n6NGjB86dO4dNmzbhhx9+QJcuXTB27Fi89dZbla6FiCqG4YaIKszf3x/dunXDe++9h3Hjxt1y3U2rVq2QmZkJtVqNyMjIKr3X3XffDY1Gg/T0dHTq1KlS++7evRvDhw/HY489BgDIz8/H2bNnLdq4ublBr9ff9jhNmzbFnj17MHToUPO2vXv3omnTppWqJzAwEMOHD8fw4cPRsWNHTJo0ieGGyIYYboioUpYsWYL27dujdevWmDVrFlq0aAEXFxccOHAAJ0+eRHR0NLp27Yq2bduiX79+ePPNN9G4cWNcuHABmzdvRr9+/cpMI5XH29sbEydOxIQJE2AwGNChQwfk5uZi79698PLywrBhw265b8OGDbF27Vr06dMHkiTh1VdfNY+kmERGRmLXrl146qmnoNFoyl3kO2nSJAwYMMC8GPq7777D2rVrLc68upPXXnsN0dHRuOeee1BcXIyNGzdWOhwRUeUw3BBRpTRo0ACpqal44403MG3aNPz111/QaDS4++67MXHiRMTFxUGSJGzevBkzZszAs88+i0uXLiEkJAQPPfQQgoODK/xec+fORVBQEBITE3H69Gn4+fmhVatWmD59+m33W7RoEZ599lm0a9cOAQEBmDJlCnJzcy3azJkzB6NHj0aDBg1QXFwMIUSZ4/Tr1w/vvPMO/vOf/+Cll15CVFQUVq1ahZiYmAp/Bjc3N0ybNg1nz56Fu7s7OnbsiC+//LLC+xNR5UmivD/RRERERDUUr3NDREREDoXhhoiIiBwKww0RERE5FIYbIiIicigMN0RERORQGG6IiIjIoTDcEBERkUNhuCEiIiKHwnBDREREDoXhhoiIiBwKww0RERE5FIYbIiIicij/D70To6uOhYmaAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_stats(statistics, ylog=False, y_max=1, view=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# They are all of the same size when the graph is full of the same color\n",
    "# plot_species(statistics, view=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "DeepLearning_Py3-10-14_TF",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
